{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "Optional: login to GenePattern\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:17.976190Z",
     "start_time": "2019-01-08T19:36:17.926802Z"
    },
    "genepattern": {
     "name": "Login",
     "server": "https://cloud.genepattern.org/gp",
     "type": "auth"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0fc172c5b22449d89b9bb6bfce38d6d1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GPAuthWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Requires GenePattern Notebook: pip install genepattern-notebook\n",
    "import gp\n",
    "import genepattern\n",
    "\n",
    "# Username and password removed for security reasons.\n",
    "genepattern.display(genepattern.session.register(\"https://cloud.genepattern.org/gp\", \"\", \"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## RNASeq workflow for Rady Children's neuro-oncology tumor board"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "This notebook contains all the routine analyses the Mesirov lab has been doing for the Rady's pediatric neuro-oncology tumor board. Its purpose is to enable another a collaborator to do these analyses quickly and reproducibly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select parameters before running the rest of the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-danger\">\n",
    "EDWIN: add option to change input and control!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "Select parameters before running the rest of the notebook.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:19.764486Z",
     "start_time": "2019-01-08T19:36:18.052291Z"
    },
    "nbtools": {
     "description": "",
     "name": "read_user_input",
     "param_values": {
      "case_id": "case21",
      "control": "custom",
      "control_expression": "NSC11.gct",
      "custom_control_expression": [
       "NSC194.gct"
      ],
      "is_medullo": "false",
      "output_var": "setup",
      "patient_dir": "'18-11222'"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7955ec2ed48c4bce965e38e7042912c2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(function_import='read_user_input', name='read_user_input', params=[{'name': 'case_id', 'label': 'cas…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "import readline # required for rpy2 extension\n",
    "%load_ext rpy2.ipython\n",
    "import os\n",
    "import sys\n",
    "import subprocess\n",
    "import warnings\n",
    "import re\n",
    "sys.path.append('/build/')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import rpy2\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from tumor_classification.medulloblastoma import classify_cavalli, classify_cho, classify_northcott\n",
    "from slides import make_medullo_classification_slide, make_discover_workflow_slide, make_exp_drug_ranking_results_slide, make_intersection_slide\n",
    "from utils import source_and_update_env_vars\n",
    "\n",
    "import pickle\n",
    "\n",
    "def rmagic_warning(\n",
    "    message,\n",
    "    category = rpy2.rinterface.RRuntimeWarning,\n",
    "    filename = '',\n",
    "    lineno = -1,\n",
    "    file=None, \n",
    "    line=None):\n",
    "    print(message)\n",
    "default_showwarning = warnings.showwarning\n",
    "\n",
    "class Bunch(object):\n",
    "  def __init__(self, adict):\n",
    "    self.__dict__.update(adict)\n",
    "\n",
    "@genepattern.build_ui(parameters={\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": False,\n",
    "    },\n",
    "    \"case_id\": {\"type\": \"text\",\n",
    "                \"description\": \"The name of the case, e.g., 'case17'\",},\n",
    "    \"patient_dir\": {\"type\": \"text\",\n",
    "                    \"description\": \"The name of the patient directory, e.g. '18-10716_tumor-normal' (quotes are required)\",},\n",
    "    \"is_medullo\": {\"type\": \"bool\",\n",
    "                   \"description\": \"Whether or not this sample has been classified as medulloblastoma\",},\n",
    "    \"control\": {\"type\": \"choice\",\n",
    "                \"description\": \"Whether or not to use a custom control\",\n",
    "                \"choices\": {\n",
    "                    \"original\": \"original\",\n",
    "                    \"custom\": \"custom\",\n",
    "                            }\n",
    "               },\n",
    "    \"custom_control_expression\": {\"type\": \"file\",\n",
    "                           \"kinds\": [\"gct\"],\n",
    "                           \"description\": \"The file (or path to the GCT file) which contains the gene expression of the custom control.\",\n",
    "                           \"default\":None},\n",
    "})\n",
    "def read_user_input(case_id, patient_dir, is_medullo=False, control='original',custom_control_expression=None):\n",
    "    # Select control for DiSCoVER and Connectivity Map\n",
    "    # Generally, if the tumor is a medulloblastoma, we use `cerebellar_stem` (comment the `neural_stem` line).\n",
    "    # And if it is any other kind of brain tumor, we use `neural_stem`.\n",
    "    \n",
    "    if control == 'original':\n",
    "        expression_control = 'cerebellar_stem' if is_medullo else 'neural_stem'\n",
    "    elif control == 'custom':\n",
    "        expression_control = 'custom_control'\n",
    "    else:\n",
    "        print('Unexpected value for variable named control, value:', control)\n",
    "        \n",
    "    if (len(custom_control_expression) is not 0) and (control is not 'custom'):\n",
    "        print(\"Reminder: if you want to use a custom control expresion, you must set control to 'custom'\")\n",
    "\n",
    "    base_dir = os.getcwd()\n",
    "    utilities_dir = '/build'\n",
    "    patients_dir = os.path.join(base_dir, 'patients')\n",
    "    in_dir = os.path.join(patients_dir, patient_dir)\n",
    "    out_dir = in_dir\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "    platform = sys.platform\n",
    "    if platform.startswith('linux'):\n",
    "        os_string = 'linux'\n",
    "    elif platform == 'darwin':\n",
    "        os_string = 'mac'\n",
    "    else:\n",
    "        raise ValueError('Platform \"{}\" not supported'.format(platform))\n",
    "\n",
    "    # RNASeq quantification\n",
    "    kallisto_dir = '/build/kallisto'\n",
    "    kallisto_path = os.path.join(kallisto_dir, 'kallisto_{}-v0.44.0/kallisto'.format(os_string))\n",
    "    transcriptome_index_path = os.path.join(kallisto_dir, 'GRCh38.ensembl.transcriptome.idx')\n",
    "    local_fastqs_dir = os.path.join(in_dir, 'fastqs')\n",
    "    os.makedirs(local_fastqs_dir, exist_ok=True)\n",
    "    patient_gexp_file = os.path.join(out_dir, 'gene_abundance.sleuth.csv')\n",
    "\n",
    "    # Medulloblastoma classification\n",
    "#     from sklearn.ensemble import RandomForestClassifier\n",
    "#     from tumor_classification.medulloblastoma import classify_cavalli, classify_cho, classify_northcott\n",
    "    medullo_classify_out_dir = os.path.join(out_dir, 'medulloblastoma_classification')\n",
    "    if not os.path.exists(medullo_classify_out_dir):\n",
    "        os.mkdir(medullo_classify_out_dir)\n",
    "    cavalli_subgroup_file = os.path.join(medullo_classify_out_dir, 'cavalli_subgroups.csv')\n",
    "    cavalli_subgroup_direct_file = os.path.join(medullo_classify_out_dir, 'cavalli_subgroups_direct.csv')\n",
    "    cavalli_subtype_file = os.path.join(medullo_classify_out_dir, 'cavalli_subtypes.csv')\n",
    "    cho_subtype_file = os.path.join(medullo_classify_out_dir, 'cho_subtypes.csv')\n",
    "    cho_subgroup_file = os.path.join(medullo_classify_out_dir, 'cho_subgroups.csv')\n",
    "    northcott_subgroup_file = os.path.join(medullo_classify_out_dir, 'northcott_subgroups.csv')\n",
    "\n",
    "    drug_suggestion_out_dir = os.path.join(out_dir, 'drug_suggestions')\n",
    "    os.makedirs(drug_suggestion_out_dir, exist_ok=True)\n",
    "\n",
    "    # DiSCoVER\n",
    "    discover_out_dir = os.path.join(drug_suggestion_out_dir, 'discover/{}'.format(expression_control))\n",
    "    os.makedirs(discover_out_dir, exist_ok=True)\n",
    "    discover_heatmap_file = os.path.join(discover_out_dir, 'ctrp.png')\n",
    "    full_discover_results_file = os.path.join(discover_out_dir, 'discover.all.csv')\n",
    "    rdrugs_discover_file = os.path.join(discover_out_dir, '{}.discover.{}.reasonable.annotated.csv'.format(case_id, expression_control))\n",
    "\n",
    "    # Connectivity Map\n",
    "    cmap_out_dir = os.path.join(drug_suggestion_out_dir, 'cmap/{}'.format(expression_control))\n",
    "    os.makedirs(cmap_out_dir, exist_ok=True)\n",
    "    cmap_all_ranked_drugs_file = os.path.join(cmap_out_dir, '{}.cmap.{}.all.csv'.format(case_id, expression_control))\n",
    "    cmap_reasonable_ranked_drugs_file = os.path.join(cmap_out_dir, '{}.cmap.{}.reasonable.annotated.csv'.format(case_id, expression_control))\n",
    "    \n",
    "    # Powerpoint for MTB\n",
    "#     from slides import make_medullo_classification_slide, make_discover_workflow_slide, make_exp_drug_ranking_results_slide, make_intersection_slide\n",
    "    mtb_ppt_file = os.path.join(out_dir, '{}.mtb_slides.pptx'.format(case_id))\n",
    "\n",
    "    # DNANexus\n",
    "    dx_source_path = os.path.join(utilities_dir, 'dx-toolkit/environment')\n",
    "    dnanexus_project = 'UW_UCSD_RNAseq_collaboration_share'\n",
    "    # Replace the contents of this file with your own DNANexus token.\n",
    "    dnanexus_token_file = os.path.join(base_dir, 'dnanexus_token.txt')\n",
    "    # To use the dx command, we must update some environment variables. \n",
    "    # From the command line, this is done with source dx-toolkit/environment, \n",
    "    # but from Python we have to use a workaround, because normally any changes \n",
    "    # to environment variables done in a subprocess are not reflected in the \n",
    "    # parent process. The workaround runs the source command in a subprocess, \n",
    "    # fetches the environment variables from the subprocess and updates those \n",
    "    # of the parent process.\n",
    "#     from utils import source_and_update_env_vars\n",
    "    source_and_update_env_vars(dx_source_path)\n",
    "    \n",
    "    out = {\"case_id\": case_id,\n",
    "                 \"patient_dir\": patient_dir,\n",
    "                 \"is_medullo\": is_medullo}\n",
    "    out['expression_control'] = expression_control\n",
    "    out['custom_control_expression'] = custom_control_expression\n",
    "    out['dnanexus_token_file'] = dnanexus_token_file\n",
    "    out['local_fastqs_dir'] = local_fastqs_dir\n",
    "    out['dnanexus_project'] = dnanexus_project\n",
    "    out['local_fastqs_dir'] = local_fastqs_dir\n",
    "    out['transcriptome_index_path'] = transcriptome_index_path\n",
    "    out['kallisto_path'] = kallisto_path\n",
    "    out['kallisto_dir'] = kallisto_dir\n",
    "    out['out_dir'] = out_dir\n",
    "    out['r_out_dir'] = out_dir.replace('\\\\',r'\\\\')\n",
    "    out['patient_gexp_file'] = patient_gexp_file\n",
    "    out['in_dir'] = in_dir\n",
    "    out['cavalli_subgroup_file'] = cavalli_subgroup_file\n",
    "    out['cavalli_subtype_file'] = cavalli_subtype_file\n",
    "    out['cavalli_subgroup_direct_file'] = cavalli_subgroup_direct_file\n",
    "    out['cho_subgroup_file'] = cho_subgroup_file\n",
    "    out['cho_subtype_file'] = cho_subtype_file\n",
    "    out['northcott_subgroup_file'] = northcott_subgroup_file\n",
    "    out['mtb_ppt_file'] = mtb_ppt_file\n",
    "    out['expression_control'] = expression_control\n",
    "    out['full_discover_results_file'] = full_discover_results_file\n",
    "    out['discover_out_dir'] = discover_out_dir\n",
    "    out['discover_heatmap_file'] = discover_heatmap_file\n",
    "    out['rdrugs_discover_file'] = rdrugs_discover_file\n",
    "    out['cmap_out_dir'] = cmap_out_dir\n",
    "    out['cmap_all_ranked_drugs_file'] = cmap_all_ranked_drugs_file\n",
    "    out['cmap_reasonable_ranked_drugs_file'] = cmap_reasonable_ranked_drugs_file\n",
    "    out['mtb_ppt_file'] = mtb_ppt_file    \n",
    "    print('Setup done!')\n",
    "    \n",
    "    return Bunch(out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"well\">\n",
    "Running all cells below this point will execute all the analyses except for one: the Connectivity Map analysis at the end of the notebook, which requires two manual steps.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download RNAseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:19.843910Z",
     "start_time": "2019-01-08T19:36:19.767486Z"
    },
    "nbtools": {
     "description": "",
     "name": "download_and_preprocess_rnaseq",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": false,
     "type": "uibuilder"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9043669d1ba14feb94bf9d9fedeffcb6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(function_import='download_and_preprocess_rnaseq', name='download_and_preprocess_rnaseq', params=[{'n…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def log(message,extra='==>'):\n",
    "    print(extra,message)\n",
    "    return\n",
    "\n",
    "\n",
    "def preprocess_rna_seq(setup):\n",
    "    if not os.path.exists(setup.transcriptome_index_path):\n",
    "        kallisto_idx_command = '{} index -i {} {}/Homo_sapiens.GRCh38.*'.format(kallisto_path, transcriptome_index_path, kallisto_dir)\n",
    "        subprocess.check_call(kallisto_idx_command, shell=True).decode('utf-8')\n",
    "    \n",
    "    fq_subdirs = [os.path.join(setup.local_fastqs_dir, subdir) for subdir in os.listdir(setup.local_fastqs_dir)]\n",
    "    ordered_fastqs = []\n",
    "    for fq_subdir in fq_subdirs:\n",
    "        if '.DS_Store' not in fq_subdir:\n",
    "            fq_files = sorted([os.path.join(fq_subdir, file) for file in os.listdir(fq_subdir) if file.endswith('fastq.gz')])\n",
    "            ordered_fastqs.extend(fq_files)\n",
    "    fastqs_str = '\" \"'.join(ordered_fastqs)\n",
    "    fastqs_str = '\"'+fastqs_str+'\"'\n",
    "    \n",
    "    command = '\"{}\" quant -i \"{}\" -o \"{}\" --bias -b 2 {}'.format(setup.kallisto_path, setup.transcriptome_index_path, setup.out_dir, fastqs_str)\n",
    "    #print(command) # if you want to execute it outside this notebook in a console\n",
    "    \n",
    "    subprocess.check_call(command, shell=True)#.decode('utf-8')    \n",
    "    return\n",
    "\n",
    "\n",
    "def run_R(command, rcript='temp_rscript.R', rlog='temp.log'):\n",
    "    print(\"Running an R command, this may take a while and will only print output at the end of the run.\")\n",
    "    with open(rcript, 'w') as f:\n",
    "        f.write(command)\n",
    "\n",
    "    subprocess.call(f'Rscript {rcript} > {rlog} 2>&1 || cat {rlog}', shell=True)\n",
    "\n",
    "    with open(rlog, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            print(line)\n",
    "    return\n",
    "\n",
    "def run_sleuth(setup):\n",
    "    warnings.showwarning = rmagic_warning # to only print the warning text, not text + returned warning object\n",
    "    from rpy2.robjects import numpy2ri\n",
    "    numpy2ri.activate()\n",
    "    r_command = f\"\"\"\n",
    "        pkg.is.installed <- function(mypkg)\n",
    "        {{\n",
    "            return(mypkg %in% rownames(installed.packages()))\n",
    "        }}\n",
    "        source(\"http://bioconductor.org/biocLite.R\")\n",
    "        if(!pkg.is.installed('biomaRt'))\n",
    "        {{\n",
    "            biocLite('biomaRt')\n",
    "        }}\n",
    "        library(\"biomaRt\")\n",
    "        mart <- biomaRt::useMart(biomart = \"ENSEMBL_MART_ENSEMBL\",\n",
    "          dataset = \"hsapiens_gene_ensembl\",\n",
    "          host = 'www.ensembl.org')\n",
    "        t2g <- biomaRt::getBM(attributes = c(\"ensembl_transcript_id\", \"external_gene_name\"), mart = mart)\n",
    "        t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id, ext_gene = external_gene_name)\n",
    "\n",
    "        if(!pkg.is.installed('rhdf5'))\n",
    "        {{\n",
    "            biocLite('rhdf5')\n",
    "        }}\n",
    "        if(!pkg.is.installed('devtools'))\n",
    "        {{\n",
    "            install.packages('devtools')\n",
    "        }}\n",
    "        if(!pkg.is.installed('sleuth'))\n",
    "        {{\n",
    "            devtools::install_github(\"pachterlab/sleuth\")\n",
    "        }}\n",
    "        library(ggplot2) # req'd by sleuth\n",
    "        library(dplyr, warn.conflicts = FALSE) # req'd by sleuth; masks some functions, but this doesn't affect us\n",
    "        library(\"sleuth\")\n",
    "        s2c = do.call(rbind.data.frame, list(c('{setup.case_id}', '{setup.case_id}', '{setup.out_dir}')))\n",
    "        colnames(s2c) <- c('sample', 'condition', 'path')\n",
    "        s2c[] <- lapply(s2c, as.character)\n",
    "        so <- sleuth_prep(s2c, ~condition, target_mapping=t2g, aggregation_column='ext_gene')\n",
    "        write.csv(so$obs_norm[, c(2,4)], '{setup.patient_gexp_file}', quote=FALSE, row.names=FALSE)\n",
    "\n",
    "        \"\"\"\n",
    "    run_R(r_command)\n",
    "    numpy2ri.deactivate()\n",
    "    # Return to default warning handling.\n",
    "    warnings.showwarning = default_showwarning\n",
    "    return\n",
    "\n",
    "\n",
    "@genepattern.build_ui(parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def download_and_preprocess_rnaseq(setup):\n",
    "    log('About to download fastqfiles from DNA Nexus. This may take a while.')\n",
    "    with open(setup.dnanexus_token_file, 'r') as f:\n",
    "        dnanexus_token = f.readline().strip()\n",
    "    login_command = 'dx login --token {} --noprojects; dx select {}'.format(dnanexus_token, setup.dnanexus_project)\n",
    "    # subprocess.check_output('ls', shell=True).decode('utf-8').strip()\n",
    "    subprocess.check_output(login_command, shell=True).decode('utf-8').strip()\n",
    "\n",
    "    find_fastq_command = 'dx find data --name \"*.fastq.gz\" --path {}:{}'.format(setup.dnanexus_project, setup.patient_dir)\n",
    "    find_fastq_return_lines = subprocess.check_output(find_fastq_command, shell=True).decode().strip().split('\\n')\n",
    "    re_string = '.*(/{}/.*\\.fastq.gz) .*'.format(setup.patient_dir)\n",
    "    fastq_path_re = re.compile(re_string)\n",
    "    remote_fastq_paths = []\n",
    "    local_fastq_subdirs = []\n",
    "\n",
    "    for line in find_fastq_return_lines:\n",
    "        search = fastq_path_re.search(line)\n",
    "        remote_fastq_path = search.group(1)\n",
    "        remote_fastq_paths.append(remote_fastq_path)\n",
    "        fastq_subdir_path = os.path.dirname(remote_fastq_path)\n",
    "        fastq_subdir = os.path.basename(fastq_subdir_path)\n",
    "        local_fastq_subdir = os.path.join(setup.local_fastqs_dir, fastq_subdir)\n",
    "        os.makedirs(local_fastq_subdir, exist_ok=True)\n",
    "        local_fastq_subdirs.append(local_fastq_subdir)\n",
    "\n",
    "    for remote_fastq_path, local_fastq_subdir in zip(remote_fastq_paths, local_fastq_subdirs):\n",
    "        download_command = 'dx download \"{}\" -o \"{}\"'.format(remote_fastq_path, local_fastq_subdir)\n",
    "        print('\\t'+download_command)\n",
    "        try:\n",
    "            a=subprocess.check_output(download_command, shell=True)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print('\\tEncountered a dx error, this likely means you already have the file indicated above.')\n",
    "            print('\\tContinuing...\\n')\n",
    "            continue\n",
    "    log('Done downloading the fastq files.')\n",
    "    log('Preprocessing RNASeq data now:')\n",
    "    log('Using kallisto to compute transcript abundance.')\n",
    "    preprocess_rna_seq(setup)\n",
    "    log('Done with tanscript abundance.')\n",
    "    log('Using sleuth to aggregate transcript abundance into gene abbundance.')\n",
    "    run_sleuth(setup)\n",
    "    patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0).T\n",
    "    patient_exp.index = [setup.case_id]\n",
    "    setup.patient_exp = patient_exp\n",
    "    patient_exp.to_csv(setup.patient_gexp_file)\n",
    "    log('Habemus Genus Expressium *release the white smoke*')\n",
    "    \n",
    "    log('Done preprocessing!')\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classify the tumor by medulloblastoma subgroup and subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what\n"
     ]
    }
   ],
   "source": [
    "print(\"what\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:19.899239Z",
     "start_time": "2019-01-08T19:36:19.849255Z"
    },
    "nbtools": {
     "description": "This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma samples are ignored.",
     "name": "classify_sample",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": false,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4eb79685dcc54e2f88b2eec05aeeac28",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(\n",
    "    description=\"This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma samples are ignored.\",\n",
    "    parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def classify_sample(setup):\n",
    "    # We have three datasets we can use to classify based on expression:\n",
    "    # - [Cavalli et al. 2017](http://www.sciencedirect.com/science/article/pii/S1535610817302015) cohort. This cohort includes 763 tumors, and was used to define 12 finer-grained subtypes nested in the 4 subgroups. Both expression and methylation data are available.\n",
    "\n",
    "    # - [Cho et al. 2011](http://www.mesirovlab.org/medulloblastoma/cho/) cohort. This paper identified two subtypes within G3 and two within G4, for a total of 6. It contains 194 tumors.\n",
    "\n",
    "    # - [Northcott et al. 2017](http://www.nature.com/nature/journal/v547/n7663/full/nature22973.html) expression data (shared by Sebastian). The labels we have for this data are of the 4 basic subgroups only. There are 223 tumors.\n",
    "\n",
    "    # When finer-grained subtypes are known, we perform the finer-grained classification first and also collapse the subtypes to the 4 basic subgroups, so as to report both subtype and subgroup probabilities. Classification is done using random forests.\n",
    "\n",
    "    # Since the patient data are from the same platform and contain the same features each time, we can use pre-fit models to classify them. The classification methods also have a fallback in case the data looks different.\n",
    "\n",
    "    # The tumor board is arranging for methylation data to be obtained from patient samples as well, since it seems it may be more informative than expression. Methylation data would also allow comparison to a large and variety collection of brain tumors, currently available through a DKFZ [web portal](https://www.molecularneuropathology.org/mnp).\n",
    "\n",
    "    if setup.is_medullo:\n",
    "        # Read in patient's gene-level RNASeq TPM data\n",
    "        patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "\n",
    "        cavalli_subgroups, cavalli_subtypes = classify_cavalli(patient_exp)\n",
    "        cavalli_subgroups.to_csv(setup.cavalli_subgroup_file)\n",
    "        cavalli_subtypes.to_csv(setup.cavalli_subtype_file)\n",
    "\n",
    "        cho_subgroups, cho_subtypes = classify_cho(patient_exp)\n",
    "        cho_subtypes.to_csv(setup.cho_subtype_file)\n",
    "        cho_subgroups.to_csv(setup.cho_subgroup_file)\n",
    "\n",
    "        northcott_subgroups = classify_northcott(patient_exp)\n",
    "        northcott_subgroups.to_csv(setup.northcott_subgroup_file)\n",
    "\n",
    "        make_medullo_classification_slide(setup.mtb_ppt_file,\n",
    "                                          setup.cavalli_subgroup_file,\n",
    "                                          setup.cavalli_subtype_file,\n",
    "                                          setup.cho_subgroup_file,\n",
    "                                          setup.cho_subtype_file,\n",
    "                                          setup.northcott_subgroup_file)\n",
    "    else:\n",
    "        print('Not a medulloblastoma sample. Nothing to do.')\n",
    "            \n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(setup, file=open(setup.case_id+'_TESTING_NSC194_NOW.p','wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "setup = pickle.load(file=open(setup.case_id+'_TESTING_NSC194_NOW.p','rb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Suggest drugs based on RNAseq data (DiSCoVER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NSC194.gct'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.custom_control_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_custom_control_exp(filename):\n",
    "    print(filename)\n",
    "    df = pd.read_table(filename, header=2, sep='\\t', index_col=0)\n",
    "    return df.drop('Description', axis=1).ix[:,0].to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:19.967914Z",
     "start_time": "2019-01-08T19:36:19.904681Z"
    },
    "nbtools": {
     "description": "Run DiSCoVER on the provided sample and control.",
     "name": "run_discover",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "46141707a85b4a068aafbf5fa5e136f7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='Run DiSCoVER on the provided sample and control.', function_import='run_discover', name…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def split_discover_dataframe(df, min_score=0):\n",
    "    # This is super inefficient for larger DataFrames, but let's worry about efficiency later\n",
    "    \n",
    "    out = pd.DataFrame(columns=['moa','GDSC','CTRP','CCLE','drug'])\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        dataset = row['drug'].split('_')[0].lower()\n",
    "        drug = row['drug'].split('_')[1].lower()\n",
    "\n",
    "        if dataset == 'gdsc':\n",
    "            out.loc[drug,'GDSC'] = row['score']\n",
    "        elif dataset == 'ctrp':\n",
    "            out.loc[drug,'CTRP'] = row['score']\n",
    "        elif dataset == 'ccle':\n",
    "            out.loc[drug,'CCLE'] = row['score']\n",
    "        out.loc[drug,'drug'] = row['drug']\n",
    "        out.loc[drug,'moa'] = row['moa']\n",
    "    \n",
    "    #     df['database'] = [drug.split('_')[0].upper() for drug in df['drug']]\n",
    "    #     df['drug'] = [drug.split('_')[1].lower() for drug in df['drug']]\n",
    "    #     df = df[df['score']>0]  # It feels inneficient to do this last, but Pandas complaints if I do it earlier.    \n",
    "    #     out = df[~df['drug'].duplicated(keep='first')].drop('score', axis=1,inplace=False)\n",
    "    #     out['Drug'] = np.unique([drug.split('_')[1].lower() for drug in df['drug']])\n",
    "    return out\n",
    "\n",
    "\n",
    "sign_to_letter = {\n",
    "    1:\"+\",\n",
    "    -1:\"-\",\n",
    "    '1.0':\"+\",\n",
    "    '-1.0':\"-\",\n",
    "    'nan':'.',\n",
    "    '0.0':'.',\n",
    "    '0':'.',\n",
    "    0:'.',\n",
    "}\n",
    "\n",
    "def supporting_evidence(row):\n",
    "    # assuming only three columns\n",
    "    return sign_to_letter[str(np.sign(row[0]))]+sign_to_letter[str(np.sign(row[1]))]+sign_to_letter[str(np.sign(row[2]))]\n",
    "\n",
    "\n",
    "def rank_drugs_discover(df):\n",
    "    df['score'] = df.drop(['moa','drug'],axis=1,inplace=False).mean(axis=1,skipna=True).round(3)\n",
    "    df['evidence'] = df.drop(['moa','score'],axis=1).apply(supporting_evidence, axis=1)\n",
    "    return df.sort_values(by=['score'],ascending=False,axis=0)\n",
    "\n",
    "\n",
    "@genepattern.build_ui(\n",
    "  description=\"Run DiSCoVER on the provided sample and control.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def run_discover(setup):\n",
    "    from rpy2.robjects import numpy2ri\n",
    "    numpy2ri.activate()\n",
    "    from drug_suggestion.expression.discover import discover_from_expression, plot_discover_from_expression\n",
    "    from drug_suggestion.expression.controls import load_control_exp\n",
    "    patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "    if setup.expression_control == 'custom_control':\n",
    "        log('Loading expression from custom control')\n",
    "        control_exp = load_custom_control_exp(setup.custom_control_expression)\n",
    "        print(control_exp)\n",
    "    else:\n",
    "        control_exp = load_control_exp(setup.expression_control)\n",
    "    log(\"About to perform DiSCoVER.\")\n",
    "    discover_results = discover_from_expression(exp=patient_exp, \n",
    "                                                control_exp=control_exp, \n",
    "                                                verbose=True)\n",
    "    log(\"DiSCoVER done!\")\n",
    "    numpy2ri.deactivate()\n",
    "    log('Saving results to file.')\n",
    "    # display(discover_results)\n",
    "    discover_results.T.sort_values(by=setup.case_id, ascending=False).to_csv(setup.full_discover_results_file)\n",
    "    log(\"Saving done!\")\n",
    "    log(\"Restricting to clinically relevant drugs.\")\n",
    "    #Not all drugs in CCLE, CTRP, and GDSC are realistic candidates for treatment. We compiled a list of medications that are FDA-approved or in late-stage clinical trials, and Dr. Wechsler-Reya curated it to include only those that are relevant for treating brain tumors. Here we limit the results to these drugs and add Dr. Wechsler-Reya's mechanism-of-action annotations. To enable comparison of drug lists, drugs from the different sources have been mapped to PubChem compound IDs (CIDs) using [PubChemPy](http://pubchempy.readthedocs.io/en/latest/).\n",
    "    from drug_suggestion.drug_annotation import subset_to_reasonable_drugs\n",
    "    from drug_suggestion.expression.discover import load_discover_drug_to_cids\n",
    "    disco2cid = load_discover_drug_to_cids()\n",
    "    reasonable_results = subset_to_reasonable_drugs(discover_results, \n",
    "                                                disco2cid, \n",
    "                                                out_prefix='discover.{}'.format(setup.expression_control), \n",
    "                                                out_dir=setup.discover_out_dir)\n",
    "    log('Done restricting to clinically relevant drugs!')\n",
    "    \n",
    "#     log('making a discover illustrative method')\n",
    "#     from drug_suggestion.expression.discover import plot_discover_from_expression\n",
    "#     plot_discover_from_expression(case_id, \n",
    "#                                   discover_results, \n",
    "#                                   exp=patient_exp,\n",
    "#                                   control_exp=control_exp,\n",
    "#                                   cl='ctrp',\n",
    "#                                   out_file=discover_heatmap_file)\n",
    "#     make_discover_workflow_slide(mtb_ppt_file, discover_heatmap_file)\n",
    "    log('Making the DiSCoVER powerpoint.')\n",
    "    rdrugs_discover = pd.read_csv(setup.rdrugs_discover_file, index_col=None)\n",
    "    df = split_discover_dataframe(df=rdrugs_discover)\n",
    "    df = rank_drugs_discover(df)\n",
    "#     df.head()\n",
    "    make_exp_drug_ranking_results_slide(setup.mtb_ppt_file, df.head(20), setup.expression_control, method='DiSCoVER')\n",
    "    log('Done making the DiSCoVER powerpoint slide!')\n",
    "    log('Savig the variables to a file.')\n",
    "    setup.discover_results = discover_results\n",
    "    setup.disco2cid = disco2cid\n",
    "    setup.control_exp = control_exp\n",
    "    setup.reasonable_results = reasonable_results\n",
    "    setup.df = df\n",
    "    pickle.dump(setup, file=open(setup.case_id+'_DISCoVER.p','wb'))\n",
    "    log('Done savig the variables to a file!')\n",
    "    \n",
    "    log('Done with all the taks in this cell. Move along.')\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Select signature genes for Connectivity Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(setup, file=open(setup.case_id+'_DISCoVER.p','wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "setup = pickle.load(file=open(setup.case_id+'_DISCoVER.p','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:20.276878Z",
     "start_time": "2019-01-08T19:36:19.970837Z"
    },
    "nbtools": {
     "description": "This function parses CMap's results.",
     "name": "make_cmap_slide",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0fe968fec2e425bbe1c7a1ad331e40a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description=\"This function parses CMap's results.\", function_import='make_cmap_slide', name='make_cm…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from drug_suggestion.expression.cmap import make_cmap_genesets, write_cmap_genesets\n",
    "from drug_suggestion.expression.cmap import read_cmap_gct, load_cmap_drug_to_cids\n",
    "from drug_suggestion.drug_annotation import subset_to_reasonable_drugs\n",
    "from drug_suggestion.expression.discover import discover_from_expression, plot_discover_from_expression\n",
    "from drug_suggestion.expression.controls import load_control_exp\n",
    "\n",
    "@genepattern.build_ui(\n",
    "  description=\"This function parses CMap's results.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def make_cmap_slide(setup):\n",
    "    log('About to parse CMap results.')\n",
    "    patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "    control_exp = load_control_exp(setup.expression_control)\n",
    "    cmap_genesets = make_cmap_genesets(patient_exp, control_exp)\n",
    "    write_cmap_genesets(cmap_genesets, setup.cmap_out_dir)\n",
    "\n",
    "    # must match path to downloaded .gct file\n",
    "    cmap_gct = os.path.join(setup.cmap_out_dir, 'cmap_result.gct')\n",
    "\n",
    "    if os.path.exists(cmap_gct):\n",
    "        cmap_ranked_drugs = read_cmap_gct(cmap_gct)\n",
    "        cmap_ranked_drugs.columns = [setup.case_id]\n",
    "        cmap_ranked_drugs.to_csv(setup.cmap_all_ranked_drugs_file)\n",
    "        cmap2cid = load_cmap_drug_to_cids()\n",
    "        cmap_reasonable = subset_to_reasonable_drugs(cmap_ranked_drugs.T, \n",
    "                                   cmap2cid,\n",
    "                                   out_prefix='cmap.{}'.format(setup.expression_control), \n",
    "                                   out_dir=setup.cmap_out_dir).sort_values(by=setup.case_id, ascending=False)\n",
    "        rdrugs_cmap = pd.read_csv(setup.cmap_reasonable_ranked_drugs_file, index_col=None)\n",
    "        make_exp_drug_ranking_results_slide(setup.mtb_ppt_file, rdrugs_cmap, setup.expression_control, method='CMap')\n",
    "        setup.rdrugs_cmap = rdrugs_cmap\n",
    "        pickle.dump(setup, file=open('patients/'+setup.patient_dir+'/drug_suggestions/'+setup.case_id+'_drug_recommendations.p','wb'))\n",
    "        log(\"done!\")\n",
    "    else:\n",
    "        log(f\"cmap_result.gct not found! (It should be present in the directiory {setup.cmap_out_dir}).\")\n",
    "        log(\"Try again if you'd like to see CMap results.\")\n",
    "        log(\"Hint, you may want to go here:\")\n",
    "        log(\"https://clue.io/l1000-query#individual\")\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DiSCoVER \"supporting information\" or \"evidence\" means in how many of the three drug databases that drug is significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-01-08T19:36:20.333913Z",
     "start_time": "2019-01-08T19:36:20.279504Z"
    },
    "nbtools": {
     "description": "This function merges the results of DiSCoVER and CMap.",
     "name": "merge_discover_and_cmap",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "94af4ab011384668890309472748b2e2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='This function merges the results of DiSCoVER and CMap.', function_import='merge_discove…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def add_cmap_to_split_df(discover,cmap):\n",
    "    df = discover.rename(index=str, columns={\"score\": \"DiSCoVER\", \"moa\":\"MoA\"},inplace=False)\n",
    "    for index, row in cmap.iterrows():\n",
    "        drug = row['drug'].lower()\n",
    "        try:\n",
    "            if str(df.loc[drug,'CMAP']) != 'nan':\n",
    "                #drug already exists and a CMAP score has been added\n",
    "                df.loc[drug,'CMAP'] = np.nanmean([row['score'], df.loc[drug,'CMAP']])\n",
    "            else:\n",
    "                # drug already exists but a CMAP score has not been added\n",
    "                df.loc[drug,'CMAP'] = row['score']\n",
    "        except KeyError:\n",
    "            # drug didn't exist therefore a CMAP score had not been added\n",
    "            df.loc[drug,'CMAP'] = row['score']    \n",
    "            df.loc[drug,'MoA'] = row['moa']\n",
    "        if str(df.loc[drug,'evidence'])=='nan': #numpy is not letting me use np.isnan()\n",
    "            df.loc[drug,'evidence'] = '...'+sign_to_letter[str(np.sign(row['score']))]\n",
    "        else:\n",
    "            df.loc[drug,'evidence'] = str(df.loc[drug,'evidence'])+sign_to_letter[str(np.sign(row['score']))]\n",
    "    \n",
    "    #update those rows which are not in cmap\n",
    "    for index, row in df.iterrows():\n",
    "        if len(row['evidence'])==3:\n",
    "            df.loc[index,'evidence']  = df.loc[index,'evidence']+'.'            \n",
    "    \n",
    "    return df[['drug','MoA','GDSC','CTRP','CCLE','DiSCoVER','CMAP','evidence']]\n",
    "\n",
    "def rank_combined_df(df):\n",
    "#     df['average'] = df.drop(['MoA','GDSC','CTRP','CCLE','support'],axis=1,inplace=False).mean(axis=1,skipna=True).round(3)\n",
    "    df.sort_values(by=['DiSCoVER'],ascending=False,axis=0,inplace=True)\n",
    "    df['DiSCoVER rank'] = range(1, len(df) + 1)\n",
    "    df.sort_values(by=['CMAP'],ascending=False,axis=0,inplace=True)\n",
    "    df['CMAP rank'] = range(1, len(df) + 1)\n",
    "    df['Average rank'] = df[['DiSCoVER rank','CMAP rank']].mean(axis=1)\n",
    "    df.sort_values(by=['Average rank'],ascending=True,axis=0,inplace=True)\n",
    "    df = df[((df['DiSCoVER']>0.001).values) & ((df['CMAP']>0.001).values)]\n",
    "#     combined_df = combined_df[((combined_df['DiSCoVER']>0.001).values) & ((combined_df['CMAP']>0.001).values)]\n",
    "\n",
    "    return df.drop(['GDSC','CTRP','CCLE'],axis=1,inplace=False)\n",
    "\n",
    "@genepattern.build_ui(\n",
    "  description=\"This function merges the results of DiSCoVER and CMap.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def merge_discover_and_cmap(setup):\n",
    "    log('Merging results from DiSCoVER and CMap.')\n",
    "    combined_df = add_cmap_to_split_df(discover=setup.df,cmap=setup.rdrugs_cmap)\n",
    "    to_slide = rank_combined_df(combined_df)\n",
    "    make_intersection_slide(setup.mtb_ppt_file, to_slide, setup.expression_control, method='DiSCoVER ∩ CMap')\n",
    "    \n",
    "    log(\"Done Merging results from DiSCoVER and CMap!\")\n",
    "    log(\"Saving combined_df and to_slide on setup variable\")\n",
    "    setup.combined_df = combined_df\n",
    "    setup.to_slide = to_slide\n",
    "    pickle.dump(setup, file=open('patients/'+setup.patient_dir+'/drug_suggestions/'+setup.case_id+'_drug_recommendations_merged.p','wb'))\n",
    "    log(\"Done done!\")\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setup.to_slide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exp_upload_command = 'dx upload --path /\"{}\"/\"{}\" \"{}\"'.format(patient_dir, os.path.basename(patient_gexp_file), patient_gexp_file)\n",
    "# subprocess.check_call(exp_upload_command, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2018-12-12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "setup = pickle.load(file=open('patients/'+setup.patient_dir+'/drug_suggestions/'+setup.case_id+'_drug_recommendations_merged.p','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>MoA</th>\n",
       "      <th>GDSC</th>\n",
       "      <th>CTRP</th>\n",
       "      <th>CCLE</th>\n",
       "      <th>DiSCoVER</th>\n",
       "      <th>CMAP</th>\n",
       "      <th>evidence</th>\n",
       "      <th>DiSCoVER rank</th>\n",
       "      <th>CMAP rank</th>\n",
       "      <th>Average rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>navitoclax</th>\n",
       "      <td>ctrp_navitoclax</td>\n",
       "      <td>Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...</td>\n",
       "      <td>0.469</td>\n",
       "      <td>0.262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.366</td>\n",
       "      <td>0.471000</td>\n",
       "      <td>++.+</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>darinaparsin</th>\n",
       "      <td>ctrp_darinaparsin</td>\n",
       "      <td>organic arsenic compound composed of dimethyla...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.207000</td>\n",
       "      <td>.+.+</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tandutinib</th>\n",
       "      <td>ctrp_tandutinib</td>\n",
       "      <td>FLT3 antagonist, also inhibits PDGFR and c-Kit,</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.247</td>\n",
       "      <td>0.605000</td>\n",
       "      <td>.+.+</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>26.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cimetidine</th>\n",
       "      <td>ctrp_cimetidine</td>\n",
       "      <td>(Tagamet) Antacid and antihistamine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.254</td>\n",
       "      <td>0.034500</td>\n",
       "      <td>.+.+.</td>\n",
       "      <td>49</td>\n",
       "      <td>13</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rucaparib</th>\n",
       "      <td>gdsc_Rucaparib</td>\n",
       "      <td>PARP inhibitor, inhibits DNA repair</td>\n",
       "      <td>0.371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.371</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>+...</td>\n",
       "      <td>10</td>\n",
       "      <td>66</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenformin</th>\n",
       "      <td>gdsc_Phenformin</td>\n",
       "      <td>Anti-diabetic AMPK Activator - mTOR Inhibitor ...</td>\n",
       "      <td>0.209</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.209</td>\n",
       "      <td>0.098000</td>\n",
       "      <td>+..+</td>\n",
       "      <td>68</td>\n",
       "      <td>8</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sitagliptin</th>\n",
       "      <td>ctrp_sitagliptin</td>\n",
       "      <td>used for Treating type 2 diabetes, works by in...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.317</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>.+..</td>\n",
       "      <td>23</td>\n",
       "      <td>61</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lenalidomide</th>\n",
       "      <td>gdsc_Lenalidomide</td>\n",
       "      <td>derivative of thalidomide, which interacts wit...</td>\n",
       "      <td>0.289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.289</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>+...</td>\n",
       "      <td>34</td>\n",
       "      <td>53</td>\n",
       "      <td>43.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyclophosphamide</th>\n",
       "      <td>ctrp_cyclophosphamide</td>\n",
       "      <td>alkylating agent, nitrogen mustard family, wor...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.324</td>\n",
       "      <td>-0.001000</td>\n",
       "      <td>.+.-</td>\n",
       "      <td>20</td>\n",
       "      <td>69</td>\n",
       "      <td>44.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>olaparib</th>\n",
       "      <td>ctrp_olaparib</td>\n",
       "      <td>PARP Inhibitor</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.298</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.326</td>\n",
       "      <td>-0.005000</td>\n",
       "      <td>++.-</td>\n",
       "      <td>18</td>\n",
       "      <td>78</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>serdemetan</th>\n",
       "      <td>gdsc_Serdemetan</td>\n",
       "      <td>(JNJ-26854165) Mdm2 antagonist, stabilizes p53</td>\n",
       "      <td>0.252</td>\n",
       "      <td>0.288</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.270</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>++..</td>\n",
       "      <td>44</td>\n",
       "      <td>54</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>veliparib</th>\n",
       "      <td>ctrp_veliparib</td>\n",
       "      <td>PARP inhibitor.</td>\n",
       "      <td>0.291</td>\n",
       "      <td>0.251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.271</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>++..</td>\n",
       "      <td>43</td>\n",
       "      <td>55</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temozolomide</th>\n",
       "      <td>ctrp_temozolomide</td>\n",
       "      <td>alkylating agent</td>\n",
       "      <td>0.376</td>\n",
       "      <td>-0.242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.018000</td>\n",
       "      <td>+-.+</td>\n",
       "      <td>84</td>\n",
       "      <td>15</td>\n",
       "      <td>49.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tretinoin</th>\n",
       "      <td>ctrp_tretinoin</td>\n",
       "      <td>Retinoid</td>\n",
       "      <td>0.523</td>\n",
       "      <td>0.322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.422</td>\n",
       "      <td>-0.042000</td>\n",
       "      <td>++..-</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>54.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abt-737</th>\n",
       "      <td>ctrp_ABT-737</td>\n",
       "      <td>Bcl-2/Bcl-xL inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.353</td>\n",
       "      <td>-0.021000</td>\n",
       "      <td>.+.-</td>\n",
       "      <td>14</td>\n",
       "      <td>97</td>\n",
       "      <td>55.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>quizartinib</th>\n",
       "      <td>ctrp_quizartinib</td>\n",
       "      <td>Flt3 inhibitor</td>\n",
       "      <td>0.476</td>\n",
       "      <td>0.232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.354</td>\n",
       "      <td>-0.022000</td>\n",
       "      <td>++.-</td>\n",
       "      <td>13</td>\n",
       "      <td>98</td>\n",
       "      <td>55.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sildenafil</th>\n",
       "      <td>ctrp_sildenafil</td>\n",
       "      <td>Viagra, used to treat erectile dysfunction and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295</td>\n",
       "      <td>-0.006000</td>\n",
       "      <td>.+.--</td>\n",
       "      <td>32</td>\n",
       "      <td>80</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prochlorperazine</th>\n",
       "      <td>ctrp_prochlorperazine</td>\n",
       "      <td>dopamine (D2) receptor antagonist, used for tr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.271</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.271</td>\n",
       "      <td>-0.002000</td>\n",
       "      <td>.+.-</td>\n",
       "      <td>42</td>\n",
       "      <td>75</td>\n",
       "      <td>58.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trifluoperazine</th>\n",
       "      <td>ctrp_trifluoperazine</td>\n",
       "      <td>antipsychotic, antiadrenergic and antidopamine...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272</td>\n",
       "      <td>-0.006000</td>\n",
       "      <td>.+.-</td>\n",
       "      <td>40</td>\n",
       "      <td>81</td>\n",
       "      <td>60.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tacrolimus</th>\n",
       "      <td>ctrp_tacrolimus</td>\n",
       "      <td>FK-506, immunosuppressive drug usedafter allog...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.309</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.309</td>\n",
       "      <td>-0.019250</td>\n",
       "      <td>.+..--</td>\n",
       "      <td>27</td>\n",
       "      <td>95</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dexamethasone</th>\n",
       "      <td>ctrp_dexamethasone</td>\n",
       "      <td>synthetic corticosteroid, used to reduce infla...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.227</td>\n",
       "      <td>-0.001219</td>\n",
       "      <td>.+.++.---</td>\n",
       "      <td>59</td>\n",
       "      <td>73</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lovastatin</th>\n",
       "      <td>ctrp_lovastatin</td>\n",
       "      <td>HMG-CoA reductase inhibitor,</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.213</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>.+..</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paclitaxel</th>\n",
       "      <td>ccle_Paclitaxel</td>\n",
       "      <td>disrupts microtubule dynamics</td>\n",
       "      <td>-0.196</td>\n",
       "      <td>-0.185</td>\n",
       "      <td>-0.211</td>\n",
       "      <td>-0.197</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>---+.</td>\n",
       "      <td>115</td>\n",
       "      <td>19</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abiraterone</th>\n",
       "      <td>ctrp_abiraterone</td>\n",
       "      <td>steroidal CYP17A1 inhibitor and androgen synth...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.230</td>\n",
       "      <td>0.289000</td>\n",
       "      <td>.-.+</td>\n",
       "      <td>134</td>\n",
       "      <td>4</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>masitinib</th>\n",
       "      <td>ctrp_masitinib</td>\n",
       "      <td>tyrosine-kinase inhibitor used in the treatmen...</td>\n",
       "      <td>0.366</td>\n",
       "      <td>-0.196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.085</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>+-..</td>\n",
       "      <td>83</td>\n",
       "      <td>56</td>\n",
       "      <td>69.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chlorambucil</th>\n",
       "      <td>ctrp_chlorambucil</td>\n",
       "      <td>alkylates and cross-links DNA, interfering wit...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.294</td>\n",
       "      <td>-0.062000</td>\n",
       "      <td>.+.-</td>\n",
       "      <td>33</td>\n",
       "      <td>107</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pazopanib</th>\n",
       "      <td>ctrp_pazopanib</td>\n",
       "      <td>KIT, FGFR, PDGFR and VEGFR inhibitor</td>\n",
       "      <td>0.356</td>\n",
       "      <td>-0.278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.039</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>+-..</td>\n",
       "      <td>85</td>\n",
       "      <td>57</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>curcumin</th>\n",
       "      <td>ctrp_curcumin</td>\n",
       "      <td>Natural Product, component of turmeric, may fu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.236</td>\n",
       "      <td>0.129000</td>\n",
       "      <td>.-.+</td>\n",
       "      <td>139</td>\n",
       "      <td>6</td>\n",
       "      <td>72.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fulvestrant</th>\n",
       "      <td>ctrp_fulvestrant</td>\n",
       "      <td>selective estrogen receptor degrader (SERD)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.217</td>\n",
       "      <td>-0.008000</td>\n",
       "      <td>.+.--</td>\n",
       "      <td>64</td>\n",
       "      <td>86</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>simvastatin</th>\n",
       "      <td>ctrp_simvastatin</td>\n",
       "      <td>HMG-CoA reductase inhibitor,</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272</td>\n",
       "      <td>-0.113000</td>\n",
       "      <td>.+.--</td>\n",
       "      <td>41</td>\n",
       "      <td>113</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pevonedistat</th>\n",
       "      <td>gdsc_Pevonedistat</td>\n",
       "      <td>(MLN4924), inhibitor of NEDD8-activating enzym...</td>\n",
       "      <td>-0.276</td>\n",
       "      <td>-0.252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>--..</td>\n",
       "      <td>156</td>\n",
       "      <td>251</td>\n",
       "      <td>203.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mitoxantrone</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Topoisomerase II inhibitor; disrupts DNA synth...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.599000</td>\n",
       "      <td>...-</td>\n",
       "      <td>262</td>\n",
       "      <td>145</td>\n",
       "      <td>203.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azd0530</th>\n",
       "      <td>ccle_AZD0530</td>\n",
       "      <td>kinase inhibitor, with selectivity against Src...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.272</td>\n",
       "      <td>-0.272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>..-.</td>\n",
       "      <td>163</td>\n",
       "      <td>252</td>\n",
       "      <td>207.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dactinomycin</th>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibits transcription by binding DNA at the t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.697500</td>\n",
       "      <td>...--</td>\n",
       "      <td>263</td>\n",
       "      <td>152</td>\n",
       "      <td>207.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>idarubicin</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Topo II inhibitor, intercalates into DNA, anth...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.725500</td>\n",
       "      <td>...--</td>\n",
       "      <td>261</td>\n",
       "      <td>154</td>\n",
       "      <td>207.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tosedostat</th>\n",
       "      <td>ctrp_tosedostat</td>\n",
       "      <td>inhibitor of the M1 family of aminopeptidases</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>164</td>\n",
       "      <td>253</td>\n",
       "      <td>208.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carmofur</th>\n",
       "      <td>NaN</td>\n",
       "      <td>derivative of fluorouracil, pyrimidine analog</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.708000</td>\n",
       "      <td>...-</td>\n",
       "      <td>264</td>\n",
       "      <td>153</td>\n",
       "      <td>208.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rapamycin</th>\n",
       "      <td>gdsc_Rapamycin</td>\n",
       "      <td>rapamycin, mTOR inhibitor</td>\n",
       "      <td>-0.278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-...</td>\n",
       "      <td>167</td>\n",
       "      <td>254</td>\n",
       "      <td>210.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>homoharringtonine</th>\n",
       "      <td>NaN</td>\n",
       "      <td>natural product, protein translation inhibitor...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.766000</td>\n",
       "      <td>...--</td>\n",
       "      <td>265</td>\n",
       "      <td>157</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>col-3</th>\n",
       "      <td>ctrp_COL-3</td>\n",
       "      <td>matrix metalloproteinase (MMP) inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>168</td>\n",
       "      <td>255</td>\n",
       "      <td>211.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tanespimycin</th>\n",
       "      <td>gdsc_Tanespimycin</td>\n",
       "      <td>HSP90 inhibitor</td>\n",
       "      <td>-0.304</td>\n",
       "      <td>-0.268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>--..</td>\n",
       "      <td>169</td>\n",
       "      <td>256</td>\n",
       "      <td>212.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pirarubicin</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Analog of doxorubicin, DNA intercalator, Topo ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.796000</td>\n",
       "      <td>...-</td>\n",
       "      <td>266</td>\n",
       "      <td>159</td>\n",
       "      <td>212.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pidorubicine</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Analog of doxorubicin, DNA intercalator, Topo ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.821000</td>\n",
       "      <td>...-</td>\n",
       "      <td>267</td>\n",
       "      <td>161</td>\n",
       "      <td>214.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>elocalcitol</th>\n",
       "      <td>ctrp_elocalcitol</td>\n",
       "      <td>vitamin D receptor agonist</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>171</td>\n",
       "      <td>257</td>\n",
       "      <td>214.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azd1480</th>\n",
       "      <td>ctrp_AZD1480</td>\n",
       "      <td>JAK2 inhibitor with some activity against JAK3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.292</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.292</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>172</td>\n",
       "      <td>258</td>\n",
       "      <td>215.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bardoxolone methyl</th>\n",
       "      <td>ctrp_bardoxolone methyl</td>\n",
       "      <td>(CDDO-Me), synthetic triterpenoid, activator o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.293</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.293</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>173</td>\n",
       "      <td>259</td>\n",
       "      <td>216.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mln-2238</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Ixazomib, Proteasome inhibitor, inhibits the c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.901000</td>\n",
       "      <td>...-</td>\n",
       "      <td>268</td>\n",
       "      <td>167</td>\n",
       "      <td>217.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vx-680</th>\n",
       "      <td>ctrp_VX-680</td>\n",
       "      <td>VX-680, pan-Aurora inhibitor, mostly against A...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>176</td>\n",
       "      <td>260</td>\n",
       "      <td>218.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd-0332991</th>\n",
       "      <td>ccle_PD-0332991</td>\n",
       "      <td>CDK4/6 inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.298</td>\n",
       "      <td>-0.298</td>\n",
       "      <td>NaN</td>\n",
       "      <td>..-.</td>\n",
       "      <td>177</td>\n",
       "      <td>261</td>\n",
       "      <td>219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azd-8055</th>\n",
       "      <td>NaN</td>\n",
       "      <td>mTOR inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.933000</td>\n",
       "      <td>...-</td>\n",
       "      <td>269</td>\n",
       "      <td>170</td>\n",
       "      <td>219.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>byl-719</th>\n",
       "      <td>ctrp_BYL-719</td>\n",
       "      <td>PI3Kα inhibitor with minimal effect on PI3Kβ/γ/δ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>179</td>\n",
       "      <td>262</td>\n",
       "      <td>220.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amsacrine</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Intercalator, inhibits Topo II and blocks both...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.936000</td>\n",
       "      <td>...-</td>\n",
       "      <td>270</td>\n",
       "      <td>171</td>\n",
       "      <td>220.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lenvatinib</th>\n",
       "      <td>ctrp_lenvatinib</td>\n",
       "      <td>VEGFR inhibitor (VEGFR1, VEGFR2 and VEGFR3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>181</td>\n",
       "      <td>263</td>\n",
       "      <td>222.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ifosfamide</th>\n",
       "      <td>ctrp_ifosfamide</td>\n",
       "      <td>alkylating agent, nitrogen mustard family, wor...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>183</td>\n",
       "      <td>264</td>\n",
       "      <td>223.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lomeguatrib</th>\n",
       "      <td>ctrp_lomeguatrib</td>\n",
       "      <td>Inhibitor of MGMT, a DNA repair protein that c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>184</td>\n",
       "      <td>265</td>\n",
       "      <td>224.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>istradefylline</th>\n",
       "      <td>ctrp_istradefylline</td>\n",
       "      <td>adenosine A2A receptor antagonist, useful in t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>185</td>\n",
       "      <td>266</td>\n",
       "      <td>225.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fluorouracil</th>\n",
       "      <td>ctrp_fluorouracil</td>\n",
       "      <td>antimetabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>187</td>\n",
       "      <td>267</td>\n",
       "      <td>227.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ibrutinib</th>\n",
       "      <td>ctrp_ibrutinib</td>\n",
       "      <td>BTK kinase inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.341</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.341</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>192</td>\n",
       "      <td>268</td>\n",
       "      <td>230.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sotrastaurin</th>\n",
       "      <td>ctrp_sotrastaurin</td>\n",
       "      <td>pan-PKC inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>196</td>\n",
       "      <td>269</td>\n",
       "      <td>232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>indisulam</th>\n",
       "      <td>ctrp_indisulam</td>\n",
       "      <td>carbonic anhydrase inibitor and  CDK inhibitor...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.-..</td>\n",
       "      <td>197</td>\n",
       "      <td>270</td>\n",
       "      <td>233.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>270 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       drug  \\\n",
       "navitoclax                  ctrp_navitoclax   \n",
       "darinaparsin              ctrp_darinaparsin   \n",
       "tandutinib                  ctrp_tandutinib   \n",
       "cimetidine                  ctrp_cimetidine   \n",
       "rucaparib                    gdsc_Rucaparib   \n",
       "phenformin                  gdsc_Phenformin   \n",
       "sitagliptin                ctrp_sitagliptin   \n",
       "lenalidomide              gdsc_Lenalidomide   \n",
       "cyclophosphamide      ctrp_cyclophosphamide   \n",
       "olaparib                      ctrp_olaparib   \n",
       "serdemetan                  gdsc_Serdemetan   \n",
       "veliparib                    ctrp_veliparib   \n",
       "temozolomide              ctrp_temozolomide   \n",
       "tretinoin                    ctrp_tretinoin   \n",
       "abt-737                        ctrp_ABT-737   \n",
       "quizartinib                ctrp_quizartinib   \n",
       "sildenafil                  ctrp_sildenafil   \n",
       "prochlorperazine      ctrp_prochlorperazine   \n",
       "trifluoperazine        ctrp_trifluoperazine   \n",
       "tacrolimus                  ctrp_tacrolimus   \n",
       "dexamethasone            ctrp_dexamethasone   \n",
       "lovastatin                  ctrp_lovastatin   \n",
       "paclitaxel                  ccle_Paclitaxel   \n",
       "abiraterone                ctrp_abiraterone   \n",
       "masitinib                    ctrp_masitinib   \n",
       "chlorambucil              ctrp_chlorambucil   \n",
       "pazopanib                    ctrp_pazopanib   \n",
       "curcumin                      ctrp_curcumin   \n",
       "fulvestrant                ctrp_fulvestrant   \n",
       "simvastatin                ctrp_simvastatin   \n",
       "...                                     ...   \n",
       "pevonedistat              gdsc_Pevonedistat   \n",
       "mitoxantrone                            NaN   \n",
       "azd0530                        ccle_AZD0530   \n",
       "dactinomycin                            NaN   \n",
       "idarubicin                              NaN   \n",
       "tosedostat                  ctrp_tosedostat   \n",
       "carmofur                                NaN   \n",
       "rapamycin                    gdsc_Rapamycin   \n",
       "homoharringtonine                       NaN   \n",
       "col-3                            ctrp_COL-3   \n",
       "tanespimycin              gdsc_Tanespimycin   \n",
       "pirarubicin                             NaN   \n",
       "pidorubicine                            NaN   \n",
       "elocalcitol                ctrp_elocalcitol   \n",
       "azd1480                        ctrp_AZD1480   \n",
       "bardoxolone methyl  ctrp_bardoxolone methyl   \n",
       "mln-2238                                NaN   \n",
       "vx-680                          ctrp_VX-680   \n",
       "pd-0332991                  ccle_PD-0332991   \n",
       "azd-8055                                NaN   \n",
       "byl-719                        ctrp_BYL-719   \n",
       "amsacrine                               NaN   \n",
       "lenvatinib                  ctrp_lenvatinib   \n",
       "ifosfamide                  ctrp_ifosfamide   \n",
       "lomeguatrib                ctrp_lomeguatrib   \n",
       "istradefylline          ctrp_istradefylline   \n",
       "fluorouracil              ctrp_fluorouracil   \n",
       "ibrutinib                    ctrp_ibrutinib   \n",
       "sotrastaurin              ctrp_sotrastaurin   \n",
       "indisulam                    ctrp_indisulam   \n",
       "\n",
       "                                                                  MoA   GDSC  \\\n",
       "navitoclax          Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...  0.469   \n",
       "darinaparsin        organic arsenic compound composed of dimethyla...    NaN   \n",
       "tandutinib           FLT3 antagonist, also inhibits PDGFR and c-Kit,     NaN   \n",
       "cimetidine                        (Tagamet) Antacid and antihistamine    NaN   \n",
       "rucaparib                         PARP inhibitor, inhibits DNA repair  0.371   \n",
       "phenformin          Anti-diabetic AMPK Activator - mTOR Inhibitor ...  0.209   \n",
       "sitagliptin         used for Treating type 2 diabetes, works by in...    NaN   \n",
       "lenalidomide        derivative of thalidomide, which interacts wit...  0.289   \n",
       "cyclophosphamide    alkylating agent, nitrogen mustard family, wor...    NaN   \n",
       "olaparib                                               PARP Inhibitor  0.353   \n",
       "serdemetan             (JNJ-26854165) Mdm2 antagonist, stabilizes p53  0.252   \n",
       "veliparib                                             PARP inhibitor.  0.291   \n",
       "temozolomide                                         alkylating agent  0.376   \n",
       "tretinoin                                                    Retinoid  0.523   \n",
       "abt-737                                        Bcl-2/Bcl-xL inhibitor    NaN   \n",
       "quizartinib                                            Flt3 inhibitor  0.476   \n",
       "sildenafil          Viagra, used to treat erectile dysfunction and...    NaN   \n",
       "prochlorperazine    dopamine (D2) receptor antagonist, used for tr...    NaN   \n",
       "trifluoperazine     antipsychotic, antiadrenergic and antidopamine...    NaN   \n",
       "tacrolimus          FK-506, immunosuppressive drug usedafter allog...    NaN   \n",
       "dexamethasone       synthetic corticosteroid, used to reduce infla...    NaN   \n",
       "lovastatin                               HMG-CoA reductase inhibitor,    NaN   \n",
       "paclitaxel                              disrupts microtubule dynamics -0.196   \n",
       "abiraterone         steroidal CYP17A1 inhibitor and androgen synth...    NaN   \n",
       "masitinib           tyrosine-kinase inhibitor used in the treatmen...  0.366   \n",
       "chlorambucil        alkylates and cross-links DNA, interfering wit...    NaN   \n",
       "pazopanib                        KIT, FGFR, PDGFR and VEGFR inhibitor  0.356   \n",
       "curcumin            Natural Product, component of turmeric, may fu...    NaN   \n",
       "fulvestrant               selective estrogen receptor degrader (SERD)    NaN   \n",
       "simvastatin                              HMG-CoA reductase inhibitor,    NaN   \n",
       "...                                                               ...    ...   \n",
       "pevonedistat        (MLN4924), inhibitor of NEDD8-activating enzym... -0.276   \n",
       "mitoxantrone        Topoisomerase II inhibitor; disrupts DNA synth...    NaN   \n",
       "azd0530             kinase inhibitor, with selectivity against Src...    NaN   \n",
       "dactinomycin        inhibits transcription by binding DNA at the t...    NaN   \n",
       "idarubicin          Topo II inhibitor, intercalates into DNA, anth...    NaN   \n",
       "tosedostat              inhibitor of the M1 family of aminopeptidases    NaN   \n",
       "carmofur                derivative of fluorouracil, pyrimidine analog    NaN   \n",
       "rapamycin                                   rapamycin, mTOR inhibitor -0.278   \n",
       "homoharringtonine   natural product, protein translation inhibitor...    NaN   \n",
       "col-3                        matrix metalloproteinase (MMP) inhibitor    NaN   \n",
       "tanespimycin                                          HSP90 inhibitor -0.304   \n",
       "pirarubicin         Analog of doxorubicin, DNA intercalator, Topo ...    NaN   \n",
       "pidorubicine        Analog of doxorubicin, DNA intercalator, Topo ...    NaN   \n",
       "elocalcitol                                vitamin D receptor agonist    NaN   \n",
       "azd1480             JAK2 inhibitor with some activity against JAK3...    NaN   \n",
       "bardoxolone methyl  (CDDO-Me), synthetic triterpenoid, activator o...    NaN   \n",
       "mln-2238            Ixazomib, Proteasome inhibitor, inhibits the c...    NaN   \n",
       "vx-680              VX-680, pan-Aurora inhibitor, mostly against A...    NaN   \n",
       "pd-0332991                                           CDK4/6 inhibitor    NaN   \n",
       "azd-8055                                               mTOR inhibitor    NaN   \n",
       "byl-719              PI3Kα inhibitor with minimal effect on PI3Kβ/γ/δ    NaN   \n",
       "amsacrine           Intercalator, inhibits Topo II and blocks both...    NaN   \n",
       "lenvatinib                VEGFR inhibitor (VEGFR1, VEGFR2 and VEGFR3)    NaN   \n",
       "ifosfamide          alkylating agent, nitrogen mustard family, wor...    NaN   \n",
       "lomeguatrib         Inhibitor of MGMT, a DNA repair protein that c...    NaN   \n",
       "istradefylline      adenosine A2A receptor antagonist, useful in t...    NaN   \n",
       "fluorouracil                                           antimetabolite    NaN   \n",
       "ibrutinib                                        BTK kinase inhibitor    NaN   \n",
       "sotrastaurin                                        pan-PKC inhibitor    NaN   \n",
       "indisulam           carbonic anhydrase inibitor and  CDK inhibitor...    NaN   \n",
       "\n",
       "                     CTRP   CCLE  DiSCoVER      CMAP   evidence  \\\n",
       "navitoclax          0.262    NaN     0.366  0.471000       ++.+   \n",
       "darinaparsin        0.328    NaN     0.328  0.207000       .+.+   \n",
       "tandutinib          0.247    NaN     0.247  0.605000       .+.+   \n",
       "cimetidine          0.254    NaN     0.254  0.034500      .+.+.   \n",
       "rucaparib             NaN    NaN     0.371 -0.000000       +...   \n",
       "phenformin            NaN    NaN     0.209  0.098000       +..+   \n",
       "sitagliptin         0.317    NaN     0.317 -0.000000       .+..   \n",
       "lenalidomide          NaN    NaN     0.289 -0.000000       +...   \n",
       "cyclophosphamide    0.324    NaN     0.324 -0.001000       .+.-   \n",
       "olaparib            0.298    NaN     0.326 -0.005000       ++.-   \n",
       "serdemetan          0.288    NaN     0.270 -0.000000       ++..   \n",
       "veliparib           0.251    NaN     0.271 -0.000000       ++..   \n",
       "temozolomide       -0.242    NaN     0.067  0.018000       +-.+   \n",
       "tretinoin           0.322    NaN     0.422 -0.042000      ++..-   \n",
       "abt-737             0.353    NaN     0.353 -0.021000       .+.-   \n",
       "quizartinib         0.232    NaN     0.354 -0.022000       ++.-   \n",
       "sildenafil          0.295    NaN     0.295 -0.006000      .+.--   \n",
       "prochlorperazine    0.271    NaN     0.271 -0.002000       .+.-   \n",
       "trifluoperazine     0.272    NaN     0.272 -0.006000       .+.-   \n",
       "tacrolimus          0.309    NaN     0.309 -0.019250     .+..--   \n",
       "dexamethasone       0.227    NaN     0.227 -0.001219  .+.++.---   \n",
       "lovastatin          0.213    NaN     0.213 -0.000000       .+..   \n",
       "paclitaxel         -0.185 -0.211    -0.197  0.007500      ---+.   \n",
       "abiraterone         -0.23    NaN    -0.230  0.289000       .-.+   \n",
       "masitinib          -0.196    NaN     0.085 -0.000000       +-..   \n",
       "chlorambucil        0.294    NaN     0.294 -0.062000       .+.-   \n",
       "pazopanib          -0.278    NaN     0.039 -0.000000       +-..   \n",
       "curcumin           -0.236    NaN    -0.236  0.129000       .-.+   \n",
       "fulvestrant         0.217    NaN     0.217 -0.008000      .+.--   \n",
       "simvastatin         0.272    NaN     0.272 -0.113000      .+.--   \n",
       "...                   ...    ...       ...       ...        ...   \n",
       "pevonedistat       -0.252    NaN    -0.264       NaN       --..   \n",
       "mitoxantrone          NaN    NaN       NaN -0.599000       ...-   \n",
       "azd0530               NaN -0.272    -0.272       NaN       ..-.   \n",
       "dactinomycin          NaN    NaN       NaN -0.697500      ...--   \n",
       "idarubicin            NaN    NaN       NaN -0.725500      ...--   \n",
       "tosedostat         -0.273    NaN    -0.273       NaN       .-..   \n",
       "carmofur              NaN    NaN       NaN -0.708000       ...-   \n",
       "rapamycin             NaN    NaN    -0.278       NaN       -...   \n",
       "homoharringtonine     NaN    NaN       NaN -0.766000      ...--   \n",
       "col-3              -0.278    NaN    -0.278       NaN       .-..   \n",
       "tanespimycin       -0.268    NaN    -0.286       NaN       --..   \n",
       "pirarubicin           NaN    NaN       NaN -0.796000       ...-   \n",
       "pidorubicine          NaN    NaN       NaN -0.821000       ...-   \n",
       "elocalcitol         -0.29    NaN    -0.290       NaN       .-..   \n",
       "azd1480            -0.292    NaN    -0.292       NaN       .-..   \n",
       "bardoxolone methyl -0.293    NaN    -0.293       NaN       .-..   \n",
       "mln-2238              NaN    NaN       NaN -0.901000       ...-   \n",
       "vx-680             -0.297    NaN    -0.297       NaN       .-..   \n",
       "pd-0332991            NaN -0.298    -0.298       NaN       ..-.   \n",
       "azd-8055              NaN    NaN       NaN -0.933000       ...-   \n",
       "byl-719            -0.301    NaN    -0.301       NaN       .-..   \n",
       "amsacrine             NaN    NaN       NaN -0.936000       ...-   \n",
       "lenvatinib         -0.304    NaN    -0.304       NaN       .-..   \n",
       "ifosfamide         -0.314    NaN    -0.314       NaN       .-..   \n",
       "lomeguatrib        -0.317    NaN    -0.317       NaN       .-..   \n",
       "istradefylline     -0.317    NaN    -0.317       NaN       .-..   \n",
       "fluorouracil       -0.321    NaN    -0.321       NaN       .-..   \n",
       "ibrutinib          -0.341    NaN    -0.341       NaN       .-..   \n",
       "sotrastaurin        -0.38    NaN    -0.380       NaN       .-..   \n",
       "indisulam          -0.405    NaN    -0.405       NaN       .-..   \n",
       "\n",
       "                    DiSCoVER rank  CMAP rank  Average rank  \n",
       "navitoclax                     11          3           7.0  \n",
       "darinaparsin                   17          5          11.0  \n",
       "tandutinib                     51          2          26.5  \n",
       "cimetidine                     49         13          31.0  \n",
       "rucaparib                      10         66          38.0  \n",
       "phenformin                     68          8          38.0  \n",
       "sitagliptin                    23         61          42.0  \n",
       "lenalidomide                   34         53          43.5  \n",
       "cyclophosphamide               20         69          44.5  \n",
       "olaparib                       18         78          48.0  \n",
       "serdemetan                     44         54          49.0  \n",
       "veliparib                      43         55          49.0  \n",
       "temozolomide                   84         15          49.5  \n",
       "tretinoin                       4        105          54.5  \n",
       "abt-737                        14         97          55.5  \n",
       "quizartinib                    13         98          55.5  \n",
       "sildenafil                     32         80          56.0  \n",
       "prochlorperazine               42         75          58.5  \n",
       "trifluoperazine                40         81          60.5  \n",
       "tacrolimus                     27         95          61.0  \n",
       "dexamethasone                  59         73          66.0  \n",
       "lovastatin                     67         67          67.0  \n",
       "paclitaxel                    115         19          67.0  \n",
       "abiraterone                   134          4          69.0  \n",
       "masitinib                      83         56          69.5  \n",
       "chlorambucil                   33        107          70.0  \n",
       "pazopanib                      85         57          71.0  \n",
       "curcumin                      139          6          72.5  \n",
       "fulvestrant                    64         86          75.0  \n",
       "simvastatin                    41        113          77.0  \n",
       "...                           ...        ...           ...  \n",
       "pevonedistat                  156        251         203.5  \n",
       "mitoxantrone                  262        145         203.5  \n",
       "azd0530                       163        252         207.5  \n",
       "dactinomycin                  263        152         207.5  \n",
       "idarubicin                    261        154         207.5  \n",
       "tosedostat                    164        253         208.5  \n",
       "carmofur                      264        153         208.5  \n",
       "rapamycin                     167        254         210.5  \n",
       "homoharringtonine             265        157         211.0  \n",
       "col-3                         168        255         211.5  \n",
       "tanespimycin                  169        256         212.5  \n",
       "pirarubicin                   266        159         212.5  \n",
       "pidorubicine                  267        161         214.0  \n",
       "elocalcitol                   171        257         214.0  \n",
       "azd1480                       172        258         215.0  \n",
       "bardoxolone methyl            173        259         216.0  \n",
       "mln-2238                      268        167         217.5  \n",
       "vx-680                        176        260         218.0  \n",
       "pd-0332991                    177        261         219.0  \n",
       "azd-8055                      269        170         219.5  \n",
       "byl-719                       179        262         220.5  \n",
       "amsacrine                     270        171         220.5  \n",
       "lenvatinib                    181        263         222.0  \n",
       "ifosfamide                    183        264         223.5  \n",
       "lomeguatrib                   184        265         224.5  \n",
       "istradefylline                185        266         225.5  \n",
       "fluorouracil                  187        267         227.0  \n",
       "ibrutinib                     192        268         230.0  \n",
       "sotrastaurin                  196        269         232.5  \n",
       "indisulam                     197        270         233.5  \n",
       "\n",
       "[270 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
