{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNASeq workflow for PDX-HTS paper\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains all the analyses..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T18:45:40.389583Z",
     "start_time": "2018-12-14T18:45:40.353410Z"
    }
   },
   "outputs": [],
   "source": [
    "# testing\n",
    "from common_imports import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T18:45:53.420435Z",
     "start_time": "2018-12-14T18:45:53.382940Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/pdx-hts/Notebooks/data/preprocessed/exp/pdx_affy_exp.csv'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PREPROC_PDX_AFFY_EXP_FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T18:45:53.710796Z",
     "start_time": "2018-12-14T18:45:53.673628Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#pdx_affy = pd.read_csv(PREPROC_PDX_AFFY_EXP_FILE, index_col=0)\n",
    "#pdx_affy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:00.143897Z",
     "start_time": "2018-12-17T15:54:00.085564Z"
    },
    "genepattern": {
     "name": "Login",
     "server": "https://genepattern.broadinstitute.org/gp",
     "type": "auth"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3ec139a3090a4aba93350572076b9f80",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GPAuthWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Requires GenePattern Notebook: pip install genepattern-notebook\n",
    "import gp\n",
    "import genepattern\n",
    "\n",
    "# Username and password removed for security reasons.\n",
    "genepattern.display(genepattern.session.register(\"https://genepattern.broadinstitute.org/gp\", \"\", \"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select parameters before running the rest of the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n",
    "Select parameters before running the rest of the notebook.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.408146Z",
     "start_time": "2018-12-17T15:54:00.213867Z"
    },
    "nbtools": {
     "description": "",
     "name": "read_user_input",
     "param_values": {
      "DNA_Nexus": "false",
      "case_id": "DMB006",
      "control": "original",
      "custom_control_expression": "None",
      "dna_nexus_bool": "false",
      "is_medullo": "false",
      "output_var": "setup",
      "patient_dir": "PDX1_dir"
     },
     "show_code": false,
     "type": "uibuilder"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "21b833bc56b84828a791285cad0c567d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(function_import='read_user_input', name='read_user_input', params=[{'name': 'case_id', 'label': 'cas…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from companion_script import *\n",
    "# # Select case\n",
    "# case_id = 'case17'\n",
    "# # This patient directory should match the directory name on DNANexus.\n",
    "# patient_dir = '18-10716_tumor-normal'\n",
    "# is_medullo = True # set False if it is another kind of brain tumor\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "import readline # required for rpy2 extension\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "\n",
    "def rmagic_warning(\n",
    "    message,\n",
    "    category = rpy2.rinterface.RRuntimeWarning,\n",
    "    filename = '',\n",
    "    lineno = -1,\n",
    "    file=None,\n",
    "    line=None):\n",
    "    print(message)\n",
    "default_showwarning = warnings.showwarning\n",
    "\n",
    "\n",
    "@genepattern.build_ui(parameters={\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": False,\n",
    "    },\n",
    "    \"case_id\": {\"type\": \"text\",\n",
    "                \"description\": \"The name of the case, e.g., 'PDX1'\",\n",
    "               \"default\":\"PDX1\"},\n",
    "    \"patient_dir\": {\"type\": \"text\",\n",
    "                    \"description\": 'For DNA Nexus downloads only. The name of the \"patient\" directory, e.g. \"18-10716_tumor-normal\" (quotes are required)',\n",
    "                    \"default\":\"PDX1_dir\"},\n",
    "    \"dna_nexus_bool\": {\"type\": \"bool\",\n",
    "                   \"description\": \"Whether or not this sample has been classified as medulloblastoma\",\n",
    "                  \"default\":False},\n",
    "    \"is_medullo\": {\"type\": \"bool\",\n",
    "                   \"description\": \"Whether or not this sample has been classified as medulloblastoma\",},\n",
    "    \"control\": {\"type\": \"choice\",\n",
    "                \"description\": \"Whether or not to use a custom control\",\n",
    "                \"choices\": {\n",
    "                    \"original\": \"original\",\n",
    "                    \"custom\": \"custom\",\n",
    "                            }\n",
    "               },\n",
    "    \"custom_control_expression\": {\"type\": \"file\",\n",
    "                           \"kinds\": [\"gct\"],\n",
    "                           \"description\": \"The file (or path to the GCT file) which contains the gene expression of the custom control.\",\n",
    "                           \"default\":None},\n",
    "})\n",
    "def read_user_input(case_id, patient_dir, dna_nexus_bool=False, is_medullo=False, control='original',custom_control_expression=None):\n",
    "    # Select control for DiSCoVER and Connectivity Map\n",
    "    # Generally, if the tumor is a medulloblastoma, we use `cerebellar_stem` (comment the `neural_stem` line).\n",
    "    # And if it is any other kind of brain tumor, we use `neural_stem`.\n",
    "    if control == 'original':\n",
    "        expression_control = 'cerebellar_stem' if is_medullo else 'neural_stem'\n",
    "    elif control == 'custom':\n",
    "        expression_control = 'custom_control'\n",
    "    else:\n",
    "        print('Unexpected value for variable named control, value:', control)\n",
    "        \n",
    "    if (len(custom_control_expression) is not 0) and (control is not 'custom'):\n",
    "        print(\"Reminder: if you want to use a custom control expresion, you must set control to 'custom'\")\n",
    "\n",
    "    base_dir = os.getcwd()\n",
    "    utilities_dir = '/build'\n",
    "    patients_dir = os.path.join(base_dir, 'patients')\n",
    "    if not dna_nexus_bool:\n",
    "        log('Setting patient_dir = case_id')\n",
    "        patient_dir = case_id\n",
    "        \n",
    "    in_dir = os.path.join(patients_dir, patient_dir)\n",
    "    \n",
    "    out_dir = in_dir\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    \n",
    "#     out['base_dir'] = base_dir\n",
    "#     out['utilities_dir'] = utilities_dir\n",
    "#     out['patients_dir'] = patients_dir\n",
    "#     \n",
    "\n",
    "    platform = sys.platform\n",
    "    if platform.startswith('linux'):\n",
    "        os_string = 'linux'\n",
    "    elif platform == 'darwin':\n",
    "        os_string = 'mac'\n",
    "    else:\n",
    "        raise ValueError('Platform \"{}\" not supported'.format(platform))\n",
    "\n",
    "    # RNASeq quantification\n",
    "    kallisto_dir = '/build/kallisto'\n",
    "    kallisto_path = os.path.join(kallisto_dir, 'kallisto_{}-v0.44.0/kallisto'.format(os_string))\n",
    "    transcriptome_index_path = os.path.join(kallisto_dir, 'GRCh38.ensembl.transcriptome.idx')\n",
    "    local_fastqs_dir = os.path.join(in_dir, 'fastqs')\n",
    "    os.makedirs(local_fastqs_dir, exist_ok=True)\n",
    "    patient_gexp_file = os.path.join(out_dir, 'gene_abundance.sleuth.csv')\n",
    "\n",
    "    # Medulloblastoma classification\n",
    "#     from sklearn.ensemble import RandomForestClassifier\n",
    "#     from tumor_classification.medulloblastoma import classify_cavalli, classify_cho, classify_northcott\n",
    "    medullo_classify_out_dir = os.path.join(out_dir, 'medulloblastoma_classification')\n",
    "    if not os.path.exists(medullo_classify_out_dir):\n",
    "        os.mkdir(medullo_classify_out_dir)\n",
    "    cavalli_subgroup_file = os.path.join(medullo_classify_out_dir, 'cavalli_subgroups.csv')\n",
    "    cavalli_subgroup_direct_file = os.path.join(medullo_classify_out_dir, 'cavalli_subgroups_direct.csv')\n",
    "    cavalli_subtype_file = os.path.join(medullo_classify_out_dir, 'cavalli_subtypes.csv')\n",
    "    cho_subtype_file = os.path.join(medullo_classify_out_dir, 'cho_subtypes.csv')\n",
    "    cho_subgroup_file = os.path.join(medullo_classify_out_dir, 'cho_subgroups.csv')\n",
    "    northcott_subgroup_file = os.path.join(medullo_classify_out_dir, 'northcott_subgroups.csv')\n",
    "\n",
    "    drug_suggestion_out_dir = os.path.join(out_dir, 'drug_suggestions')\n",
    "    os.makedirs(drug_suggestion_out_dir, exist_ok=True)\n",
    "\n",
    "    # DiSCoVER\n",
    "    discover_out_dir = os.path.join(drug_suggestion_out_dir, 'discover/{}'.format(expression_control))\n",
    "    os.makedirs(discover_out_dir, exist_ok=True)\n",
    "    discover_heatmap_file = os.path.join(discover_out_dir, 'ctrp.png')\n",
    "    full_discover_results_file = os.path.join(discover_out_dir, 'discover.all.csv')\n",
    "    rdrugs_discover_file = os.path.join(discover_out_dir, '{}.discover.{}.reasonable.annotated.csv'.format(case_id, expression_control))\n",
    "\n",
    "    # Connectivity Map\n",
    "    cmap_out_dir = os.path.join(drug_suggestion_out_dir, 'cmap/{}'.format(expression_control))\n",
    "    os.makedirs(cmap_out_dir, exist_ok=True)\n",
    "    cmap_all_ranked_drugs_file = os.path.join(cmap_out_dir, '{}.cmap.{}.all.csv'.format(case_id, expression_control))\n",
    "    cmap_reasonable_ranked_drugs_file = os.path.join(cmap_out_dir, '{}.cmap.{}.reasonable.annotated.csv'.format(case_id, expression_control))\n",
    "    \n",
    "    # Powerpoint for MTB\n",
    "#     from slides import make_medullo_classification_slide, make_discover_workflow_slide, make_exp_drug_ranking_results_slide, make_intersection_slide\n",
    "    mtb_ppt_file = os.path.join(out_dir, '{}.mtb_slides.pptx'.format(case_id))\n",
    "\n",
    "    # DNANexus\n",
    "    dx_source_path = os.path.join(utilities_dir, 'dx-toolkit/environment')\n",
    "    dnanexus_project = 'UW_UCSD_RNAseq_collaboration_share'\n",
    "    # Replace the contents of this file with your own DNANexus token.\n",
    "    dnanexus_token_file = os.path.join(base_dir, 'dnanexus_token.txt')\n",
    "    # To use the dx command, we must update some environment variables. \n",
    "    # From the command line, this is done with source dx-toolkit/environment, \n",
    "    # but from Python we have to use a workaround, because normally any changes \n",
    "    # to environment variables done in a subprocess are not reflected in the \n",
    "    # parent process. The workaround runs the source command in a subprocess, \n",
    "    # fetches the environment variables from the subprocess and updates those \n",
    "    # of the parent process.\n",
    "#     from utils import source_and_update_env_vars\n",
    "    source_and_update_env_vars(dx_source_path)    \n",
    "    out = {\"case_id\": case_id,\n",
    "                 \"patient_dir\": patient_dir,\n",
    "                 \"is_medullo\": is_medullo}\n",
    "    out['dna_nexus_bool']=dna_nexus_bool\n",
    "    out['expression_control'] = expression_control\n",
    "    out['custom_control_expression'] = custom_control_expression\n",
    "    out['dnanexus_token_file'] = dnanexus_token_file\n",
    "    out['local_fastqs_dir'] = local_fastqs_dir\n",
    "    out['dnanexus_project'] = dnanexus_project\n",
    "    out['local_fastqs_dir'] = local_fastqs_dir\n",
    "    out['transcriptome_index_path'] = transcriptome_index_path\n",
    "    out['kallisto_path'] = kallisto_path\n",
    "    out['kallisto_dir'] = kallisto_dir\n",
    "    out['out_dir'] = out_dir\n",
    "    out['r_out_dir'] = out_dir.replace('\\\\',r'\\\\')\n",
    "    out['patient_gexp_file'] = patient_gexp_file\n",
    "    out['in_dir'] = in_dir\n",
    "    out['cavalli_subgroup_file'] = cavalli_subgroup_file\n",
    "    out['cavalli_subtype_file'] = cavalli_subtype_file\n",
    "    out['cavalli_subgroup_direct_file'] = cavalli_subgroup_direct_file\n",
    "    out['cho_subgroup_file'] = cho_subgroup_file\n",
    "    out['cho_subtype_file'] = cho_subtype_file\n",
    "    out['northcott_subgroup_file'] = northcott_subgroup_file\n",
    "    out['mtb_ppt_file'] = mtb_ppt_file\n",
    "    out['expression_control'] = expression_control\n",
    "    out['full_discover_results_file'] = full_discover_results_file\n",
    "    out['discover_out_dir'] = discover_out_dir\n",
    "    out['discover_heatmap_file'] = discover_heatmap_file\n",
    "    out['rdrugs_discover_file'] = rdrugs_discover_file\n",
    "    out['cmap_out_dir'] = cmap_out_dir\n",
    "    out['cmap_all_ranked_drugs_file'] = cmap_all_ranked_drugs_file\n",
    "    out['cmap_reasonable_ranked_drugs_file'] = cmap_reasonable_ranked_drugs_file\n",
    "    out['mtb_ppt_file'] = mtb_ppt_file\n",
    "    out['out_dir'] = os.path.join(patients_dir, case_id)\n",
    "\n",
    "    if not os.path.exists(out['out_dir']):\n",
    "        os.mkdir(out['out_dir'])\n",
    "    print('Setup done!')\n",
    "    pickle.dump(out, file=open(os.path.join(out['out_dir'], case_id+'_backup1_input.p'),'wb'))\n",
    "    return Bunch(out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"well\">\n",
    "Running all cells below this point will execute all the analyses except for one: the Connectivity Map analysis at the end of the notebook, which requires two manual steps.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download RNAseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.509507Z",
     "start_time": "2018-12-17T15:54:04.410786Z"
    },
    "nbtools": {
     "description": "",
     "name": "download_and_preprocess_rnaseq",
     "param_values": {
      "input_expression_dir": "None",
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": false,
     "type": "uibuilder"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2e6d1d0936b94b87a2cf4eb8a52b33e6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(function_import='download_and_preprocess_rnaseq', name='download_and_preprocess_rnaseq', params=[{'n…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"input_expression_dir\":{\"hide\":True},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def download_and_preprocess_rnaseq(setup, input_expression_dir = None):\n",
    "    setup.input_expression_dir = f'/pdx-hts/Notebooks/data/preprocessed/exp/{setup.case_id}.csv'\n",
    "    if setup.dna_nexus_bool:\n",
    "        if input_expression_dir is not None:\n",
    "            log(f\"input_expression_dir has a value ({setup.input_expression_dir}), and it will be ignored!\")\n",
    "        log('About to download fastqfiles from DNA Nexus. This may take a while.')\n",
    "        with open(setup.dnanexus_token_file, 'r') as f:\n",
    "            dnanexus_token = f.readline().strip()\n",
    "        login_command = 'dx login --token {} --noprojects; dx select {}'.format(dnanexus_token, setup.dnanexus_project)\n",
    "        # subprocess.check_output('ls', shell=True).decode('utf-8').strip()\n",
    "        subprocess.check_output(login_command, shell=True).decode('utf-8').strip()\n",
    "\n",
    "        find_fastq_command = 'dx find data --name \"*.fastq.gz\" --path {}:{}'.format(setup.dnanexus_project, setup.patient_dir)\n",
    "        find_fastq_return_lines = subprocess.check_output(find_fastq_command, shell=True).decode().strip().split('\\n')\n",
    "        re_string = '.*(/{}/.*\\.fastq.gz) .*'.format(setup.patient_dir)\n",
    "        fastq_path_re = re.compile(re_string)\n",
    "        remote_fastq_paths = []\n",
    "        local_fastq_subdirs = []\n",
    "\n",
    "        for line in find_fastq_return_lines:\n",
    "            search = fastq_path_re.search(line)\n",
    "            remote_fastq_path = search.group(1)\n",
    "            remote_fastq_paths.append(remote_fastq_path)\n",
    "            fastq_subdir_path = os.path.dirname(remote_fastq_path)\n",
    "            fastq_subdir = os.path.basename(fastq_subdir_path)\n",
    "            local_fastq_subdir = os.path.join(setup.local_fastqs_dir, fastq_subdir)\n",
    "            os.makedirs(local_fastq_subdir, exist_ok=True)\n",
    "            local_fastq_subdirs.append(local_fastq_subdir)\n",
    "\n",
    "        for remote_fastq_path, local_fastq_subdir in zip(remote_fastq_paths, local_fastq_subdirs):\n",
    "            download_command = 'dx download \"{}\" -o \"{}\"'.format(remote_fastq_path, local_fastq_subdir)\n",
    "            print('\\t'+download_command)\n",
    "            try:\n",
    "                a=subprocess.check_output(download_command, shell=True)\n",
    "            except subprocess.CalledProcessError as e:\n",
    "                print('\\tEncountered a dx error, this likely means you already have the file indicated above.')\n",
    "                print('\\tContinuing...\\n')\n",
    "                continue\n",
    "        log('Done downloading the fastq files.')\n",
    "        log('Preprocessing RNASeq data now:')\n",
    "        log('Using kallisto to compute transcript abundance.')\n",
    "        preprocess_rna_seq(setup)\n",
    "        log('Done with tanscript abundance.')\n",
    "        log('Using sleuth to aggregate transcript abundance into gene abbundance.')\n",
    "        run_sleuth(setup)\n",
    "        patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0).T\n",
    "        patient_exp.index = [setup.case_id]\n",
    "        setup.patient_exp = patient_exp\n",
    "        patient_exp.to_csv(setup.patient_gexp_file)\n",
    "        log('Habemus Genus Expressium *release the white smoke*')\n",
    "    else:\n",
    "        log(f'Checking if local file ({setup.input_expression_dir}) exist.')\n",
    "        if os.path.isfile(setup.input_expression_dir):\n",
    "            df = pd.read_csv(setup.input_expression_dir, index_col=0)\n",
    "            setup.expression_input = df\n",
    "            log(\"This file containes the expression of the PDXs. Printing dataframe's info:\")\n",
    "            log(setup.expression_input.info())\n",
    "        else:\n",
    "            log('File could not be located please check and run again.')\n",
    "        patient_exp = df\n",
    "        setup.patient_exp = patient_exp\n",
    "        patient_exp.to_csv(setup.patient_gexp_file)\n",
    "        log(f'File {setup.patient_gexp_file} saved successfully')\n",
    "    \n",
    "    pickle.dump(setup, file=open(os.path.join(setup.out_dir, setup.case_id+'_backup2_download.p'),'wb'))\n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup2_download.p','wb'))\n",
    "    log('Done preprocessing!')\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classify the tumor by medulloblastoma subgroup and subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.555571Z",
     "start_time": "2018-12-17T15:54:04.512753Z"
    },
    "nbtools": {
     "description": "This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma samples are ignored.",
     "name": "classify_sample",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": false,
     "type": "uibuilder"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "22551d0a3c3440bf9fede929996e7eec",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(\n",
    "    description=\"This function classifies a medulloblastoma sample into a subgroup. Non-medulloblastoma samples are ignored.\",\n",
    "    parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def classify_sample(setup):\n",
    "    # We have three datasets we can use to classify based on expression:\n",
    "    # - [Cavalli et al. 2017](http://www.sciencedirect.com/science/article/pii/S1535610817302015) cohort. This cohort includes 763 tumors, and was used to define 12 finer-grained subtypes nested in the 4 subgroups. Both expression and methylation data are available.\n",
    "\n",
    "    # - [Cho et al. 2011](http://www.mesirovlab.org/medulloblastoma/cho/) cohort. This paper identified two subtypes within G3 and two within G4, for a total of 6. It contains 194 tumors.\n",
    "\n",
    "    # - [Northcott et al. 2017](http://www.nature.com/nature/journal/v547/n7663/full/nature22973.html) expression data (shared by Sebastian). The labels we have for this data are of the 4 basic subgroups only. There are 223 tumors.\n",
    "\n",
    "    # When finer-grained subtypes are known, we perform the finer-grained classification first and also collapse the subtypes to the 4 basic subgroups, so as to report both subtype and subgroup probabilities. Classification is done using random forests.\n",
    "\n",
    "    # Since the patient data are from the same platform and contain the same features each time, we can use pre-fit models to classify them. The classification methods also have a fallback in case the data looks different.\n",
    "\n",
    "    # The tumor board is arranging for methylation data to be obtained from patient samples as well, since it seems it may be more informative than expression. Methylation data would also allow comparison to a large and variety collection of brain tumors, currently available through a DKFZ [web portal](https://www.molecularneuropathology.org/mnp).\n",
    "\n",
    "    if setup.is_medullo:\n",
    "        # Read in patient's gene-level RNASeq TPM data\n",
    "        patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "\n",
    "        cavalli_subgroups, cavalli_subtypes = classify_cavalli(patient_exp)\n",
    "        cavalli_subgroups.to_csv(setup.cavalli_subgroup_file)\n",
    "        cavalli_subtypes.to_csv(setup.cavalli_subtype_file)\n",
    "\n",
    "        cho_subgroups, cho_subtypes = classify_cho(patient_exp)\n",
    "        cho_subtypes.to_csv(setup.cho_subtype_file)\n",
    "        cho_subgroups.to_csv(setup.cho_subgroup_file)\n",
    "\n",
    "        northcott_subgroups = classify_northcott(patient_exp)\n",
    "        northcott_subgroups.to_csv(setup.northcott_subgroup_file)\n",
    "\n",
    "        make_medullo_classification_slide(setup.mtb_ppt_file,\n",
    "                                          setup.cavalli_subgroup_file,\n",
    "                                          setup.cavalli_subtype_file,\n",
    "                                          setup.cho_subgroup_file,\n",
    "                                          setup.cho_subtype_file,\n",
    "                                          setup.northcott_subgroup_file)\n",
    "        log('Done! Move along')\n",
    "    else:\n",
    "        log('This is not medulloblastoma. Nothing to do here. Move along')\n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup3_classify.p','wb'))\n",
    "    pickle.dump(setup, file=open(os.path.join(setup.out_dir, setup.case_id+'_backup3_classify.p'),'wb'))\n",
    "    \n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T20:56:56.199159Z",
     "start_time": "2018-12-14T20:56:56.144842Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AACSP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZUP1</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DMB006</th>\n",
       "      <td>3.359795</td>\n",
       "      <td>3.317396</td>\n",
       "      <td>2.982897</td>\n",
       "      <td>3.272958</td>\n",
       "      <td>5.974374</td>\n",
       "      <td>2.700876</td>\n",
       "      <td>3.623813</td>\n",
       "      <td>4.469528</td>\n",
       "      <td>8.874965</td>\n",
       "      <td>3.177392</td>\n",
       "      <td>...</td>\n",
       "      <td>7.441817</td>\n",
       "      <td>8.600564</td>\n",
       "      <td>6.104677</td>\n",
       "      <td>5.542006</td>\n",
       "      <td>6.384078</td>\n",
       "      <td>3.845576</td>\n",
       "      <td>9.761523</td>\n",
       "      <td>4.290367</td>\n",
       "      <td>5.395606</td>\n",
       "      <td>8.178586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 16247 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A1BG  A1BG-AS1      A1CF       A2M   A2M-AS1     A2ML1     A4GNT  \\\n",
       "DMB006  3.359795  3.317396  2.982897  3.272958  5.974374  2.700876  3.623813   \n",
       "\n",
       "            AAAS      AACS    AACSP1    ...         ZUP1    ZWILCH      ZXDA  \\\n",
       "DMB006  4.469528  8.874965  3.177392    ...     7.441817  8.600564  6.104677   \n",
       "\n",
       "            ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "DMB006  5.542006  6.384078  3.845576  9.761523  4.290367  5.395606  8.178586  \n",
       "\n",
       "[1 rows x 16247 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.expression_input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Suggest drugs based on RNAseq data (DiSCoVER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.609745Z",
     "start_time": "2018-12-17T15:54:04.560008Z"
    },
    "nbtools": {
     "description": "Run DiSCoVER on the provided sample and control.",
     "name": "run_discover",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1bb5de1e79a440a6aac5aa7140158ad9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='Run DiSCoVER on the provided sample and control.', function_import='run_discover', name…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(\n",
    "  description=\"Run DiSCoVER on the provided sample and control.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def run_discover(setup):\n",
    "    from rpy2.robjects import numpy2ri\n",
    "    numpy2ri.activate()\n",
    "    from discover import discover_from_expression, plot_discover_from_expression\n",
    "    from drug_suggestion.expression.controls import load_control_exp\n",
    "    patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "    control_exp = load_control_exp(setup.expression_control)\n",
    "    log(\"About to perform DiSCoVER.\")\n",
    "    discover_results = discover_from_expression(exp=patient_exp, \n",
    "                                                control_exp=control_exp, \n",
    "                                                verbose=False)\n",
    "    setup.raw_discover_results = discover_results\n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup4-0_DISCoVER.p','wb'))\n",
    "#     print(discover_results)\n",
    "    # move some files created by DiSCoVER\n",
    "#     for cl_name in ['ccle','ctrp','gdsc']:\n",
    "#         os.rename(f\"COSMIC_cell_lines_IDs_and_types_{cl_name}.csv\", os.path.join(setup.out_dir, f\"COSMIC_cell_lines_IDs_and_types_{cl_name}.csv\"))\n",
    "    \n",
    "    log(\"DiSCoVER done!\")\n",
    "    numpy2ri.deactivate()\n",
    "    log('Saving results to file.')\n",
    "    # display(discover_results)\n",
    "    discover_results.T.sort_values(by=setup.case_id, ascending=False).to_csv(setup.full_discover_results_file)\n",
    "    log(\"Saving done!\")\n",
    "    log(\"NOT Restricting to clinically relevant drugs.\")\n",
    "    #Not all drugs in CCLE, CTRP, and GDSC are realistic candidates for treatment. We compiled a list of medications that are FDA-approved or in late-stage clinical trials, and Dr. Wechsler-Reya curated it to include only those that are relevant for treating brain tumors. Here we limit the results to these drugs and add Dr. Wechsler-Reya's mechanism-of-action annotations. To enable comparison of drug lists, drugs from the different sources have been mapped to PubChem compound IDs (CIDs) using [PubChemPy](http://pubchempy.readthedocs.io/en/latest/).\n",
    "    from drug_suggestion.drug_annotation import subset_to_reasonable_drugs\n",
    "    from drug_suggestion.expression.discover import load_discover_drug_to_cids\n",
    "    disco2cid = load_discover_drug_to_cids()\n",
    "    reasonable_results = subset_to_reasonable_drugs(discover_results, \n",
    "                                                disco2cid, \n",
    "                                                out_prefix='discover.{}'.format(setup.expression_control), \n",
    "                                                out_dir=setup.discover_out_dir)\n",
    "    # This will override the file setup.rdrugs_discover_file\n",
    "    all_drugs = format_drugs(discover_results, \n",
    "                                disco2cid, \n",
    "                                out_prefix='discover.{}'.format(setup.expression_control), \n",
    "                                out_dir=setup.discover_out_dir)\n",
    "    log('Done restricting to clinically relevant drugs!')\n",
    "    \n",
    "##     log('making a discover illustrative method')\n",
    "##     from drug_suggestion.expression.discover import plot_discover_from_expression\n",
    "##     plot_discover_from_expression(case_id, \n",
    "##                                   discover_results, \n",
    "##                                   exp=patient_exp,\n",
    "##                                   control_exp=control_exp,\n",
    "##                                   cl='ctrp',\n",
    "##                                   out_file=discover_heatmap_file)\n",
    "##     make_discover_workflow_slide(mtb_ppt_file, discover_heatmap_file)\n",
    "    log('Making the DiSCoVER powerpoint.')\n",
    "    rdrugs_discover = pd.read_csv(setup.rdrugs_discover_file, index_col=None)\n",
    "\n",
    "    \n",
    "    # Using all of the drugs\n",
    "    log('Using all of the drugs')\n",
    "    df = split_discover_dataframe(df=rdrugs_discover)\n",
    "    df = rank_drugs_discover(df)\n",
    "#     df.head()\n",
    "    make_exp_drug_ranking_results_slide(setup.mtb_ppt_file, df.head(20), setup.expression_control, method='DiSCoVER')\n",
    "    log('Done making the DiSCoVER powerpoint slide!')\n",
    "    log('Savig the variables to a file.')\n",
    "    setup.discover_results = discover_results\n",
    "    setup.disco2cid = disco2cid\n",
    "    setup.control_exp = control_exp\n",
    "    setup.reasonable_results = reasonable_results\n",
    "    setup.df = df\n",
    "    log('Saving the formatted results of DiSCoVER to a csv')\n",
    "    df.to_csv(os.path.join(setup.out_dir, setup.case_id+'_formated_DISCoVER_results.csv'))\n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup4_DISCoVER.p','wb'))\n",
    "    pickle.dump(setup, file=open(os.path.join(setup.out_dir, setup.case_id+'_backup4_DISCoVER.p'),'wb'))\n",
    "#     pickle.dump(setup, file=open(setup.case_id+'_DISCoVER.p','wb'))\n",
    "    log('Done savig the variables to a file!')\n",
    "    \n",
    "    log('Done with all the taks in this cell. Move along.')\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T21:49:56.730165Z",
     "start_time": "2018-12-14T21:49:56.679742Z"
    }
   },
   "outputs": [],
   "source": [
    "setup.df.to_csv(os.path.join(setup.out_dir, setup.case_id+'_formated_DISCoVER_results.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T20:58:51.086871Z",
     "start_time": "2018-12-14T20:58:51.022092Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ctrp_ML311</th>\n",
       "      <th>ctrp_zebularine</th>\n",
       "      <th>ctrp_BRD-A02303741</th>\n",
       "      <th>ctrp_importazole</th>\n",
       "      <th>ctrp_SR8278</th>\n",
       "      <th>ctrp_elocalcitol</th>\n",
       "      <th>ctrp_BRD-A05715709</th>\n",
       "      <th>ctrp_docetaxel</th>\n",
       "      <th>ctrp_cyclophosphamide</th>\n",
       "      <th>ctrp_3-Cl-AHPC</th>\n",
       "      <th>...</th>\n",
       "      <th>ccle_PHA-665752</th>\n",
       "      <th>ccle_PLX4720</th>\n",
       "      <th>ccle_Paclitaxel</th>\n",
       "      <th>ccle_Panobinostat</th>\n",
       "      <th>ccle_RAF265</th>\n",
       "      <th>ccle_Sorafenib</th>\n",
       "      <th>ccle_TAE684</th>\n",
       "      <th>ccle_TKI258</th>\n",
       "      <th>ccle_Topotecan</th>\n",
       "      <th>ccle_ZD-6474</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DMB006</th>\n",
       "      <td>0.24922</td>\n",
       "      <td>-0.368372</td>\n",
       "      <td>0.224963</td>\n",
       "      <td>0.284399</td>\n",
       "      <td>-0.359116</td>\n",
       "      <td>0.345501</td>\n",
       "      <td>-0.279495</td>\n",
       "      <td>0.216308</td>\n",
       "      <td>0.289871</td>\n",
       "      <td>0.244442</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.262857</td>\n",
       "      <td>-0.29992</td>\n",
       "      <td>-0.21247</td>\n",
       "      <td>-0.277326</td>\n",
       "      <td>-0.207103</td>\n",
       "      <td>-0.195914</td>\n",
       "      <td>-0.231196</td>\n",
       "      <td>-0.245791</td>\n",
       "      <td>0.21383</td>\n",
       "      <td>-0.257422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 755 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ctrp_ML311  ctrp_zebularine  ctrp_BRD-A02303741  ctrp_importazole  \\\n",
       "DMB006     0.24922        -0.368372            0.224963          0.284399   \n",
       "\n",
       "        ctrp_SR8278  ctrp_elocalcitol  ctrp_BRD-A05715709  ctrp_docetaxel  \\\n",
       "DMB006    -0.359116          0.345501           -0.279495        0.216308   \n",
       "\n",
       "        ctrp_cyclophosphamide  ctrp_3-Cl-AHPC      ...       ccle_PHA-665752  \\\n",
       "DMB006               0.289871        0.244442      ...             -0.262857   \n",
       "\n",
       "        ccle_PLX4720  ccle_Paclitaxel  ccle_Panobinostat  ccle_RAF265  \\\n",
       "DMB006      -0.29992         -0.21247          -0.277326    -0.207103   \n",
       "\n",
       "        ccle_Sorafenib  ccle_TAE684  ccle_TKI258  ccle_Topotecan  ccle_ZD-6474  \n",
       "DMB006       -0.195914    -0.231196    -0.245791         0.21383     -0.257422  \n",
       "\n",
       "[1 rows x 755 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.raw_discover_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T20:59:00.900003Z",
     "start_time": "2018-12-14T20:59:00.841930Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ctrp_ML311</th>\n",
       "      <th>ctrp_zebularine</th>\n",
       "      <th>ctrp_BRD-A02303741</th>\n",
       "      <th>ctrp_importazole</th>\n",
       "      <th>ctrp_SR8278</th>\n",
       "      <th>ctrp_elocalcitol</th>\n",
       "      <th>ctrp_BRD-A05715709</th>\n",
       "      <th>ctrp_docetaxel</th>\n",
       "      <th>ctrp_cyclophosphamide</th>\n",
       "      <th>ctrp_3-Cl-AHPC</th>\n",
       "      <th>...</th>\n",
       "      <th>ccle_PHA-665752</th>\n",
       "      <th>ccle_PLX4720</th>\n",
       "      <th>ccle_Paclitaxel</th>\n",
       "      <th>ccle_Panobinostat</th>\n",
       "      <th>ccle_RAF265</th>\n",
       "      <th>ccle_Sorafenib</th>\n",
       "      <th>ccle_TAE684</th>\n",
       "      <th>ccle_TKI258</th>\n",
       "      <th>ccle_Topotecan</th>\n",
       "      <th>ccle_ZD-6474</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DMB006</th>\n",
       "      <td>0.24922</td>\n",
       "      <td>-0.368372</td>\n",
       "      <td>0.224963</td>\n",
       "      <td>0.284399</td>\n",
       "      <td>-0.359116</td>\n",
       "      <td>0.345501</td>\n",
       "      <td>-0.279495</td>\n",
       "      <td>0.216308</td>\n",
       "      <td>0.289871</td>\n",
       "      <td>0.244442</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.262857</td>\n",
       "      <td>-0.29992</td>\n",
       "      <td>-0.21247</td>\n",
       "      <td>-0.277326</td>\n",
       "      <td>-0.207103</td>\n",
       "      <td>-0.195914</td>\n",
       "      <td>-0.231196</td>\n",
       "      <td>-0.245791</td>\n",
       "      <td>0.21383</td>\n",
       "      <td>-0.257422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 755 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ctrp_ML311  ctrp_zebularine  ctrp_BRD-A02303741  ctrp_importazole  \\\n",
       "DMB006     0.24922        -0.368372            0.224963          0.284399   \n",
       "\n",
       "        ctrp_SR8278  ctrp_elocalcitol  ctrp_BRD-A05715709  ctrp_docetaxel  \\\n",
       "DMB006    -0.359116          0.345501           -0.279495        0.216308   \n",
       "\n",
       "        ctrp_cyclophosphamide  ctrp_3-Cl-AHPC      ...       ccle_PHA-665752  \\\n",
       "DMB006               0.289871        0.244442      ...             -0.262857   \n",
       "\n",
       "        ccle_PLX4720  ccle_Paclitaxel  ccle_Panobinostat  ccle_RAF265  \\\n",
       "DMB006      -0.29992         -0.21247          -0.277326    -0.207103   \n",
       "\n",
       "        ccle_Sorafenib  ccle_TAE684  ccle_TKI258  ccle_Topotecan  ccle_ZD-6474  \n",
       "DMB006       -0.195914    -0.231196    -0.245791         0.21383     -0.257422  \n",
       "\n",
       "[1 rows x 755 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.discover_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T21:22:17.067566Z",
     "start_time": "2018-12-14T21:22:16.695525Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/pandas/util/_decorators.py:118: FutureWarning: The `sheetname` keyword is deprecated, use `sheet_name` instead\n",
      "  return func(*args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DMB006</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ctrp_elocalcitol</th>\n",
       "      <td>0.345501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_RAF265</th>\n",
       "      <td>0.211422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Vinblastine</th>\n",
       "      <td>-0.282946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Tamoxifen</th>\n",
       "      <td>-0.344953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_temozolomide</th>\n",
       "      <td>0.273595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ccle_Paclitaxel</th>\n",
       "      <td>-0.212470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_quizartinib</th>\n",
       "      <td>0.269583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Sorafenib</th>\n",
       "      <td>-0.487374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_BI-2536</th>\n",
       "      <td>-0.298340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_daporinad</th>\n",
       "      <td>0.248200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_MGCD-265</th>\n",
       "      <td>0.309477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Bortezomib</th>\n",
       "      <td>-0.373787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_indisulam</th>\n",
       "      <td>0.390994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_sunitinib</th>\n",
       "      <td>0.283824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Ponatinib</th>\n",
       "      <td>-0.337840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Methotrexate</th>\n",
       "      <td>-0.264438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_RO4929097</th>\n",
       "      <td>0.290939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_TG-101348</th>\n",
       "      <td>0.238428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Axitinib</th>\n",
       "      <td>-0.489878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_saracatinib</th>\n",
       "      <td>-0.252630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_linsitinib</th>\n",
       "      <td>0.449036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_etoposide</th>\n",
       "      <td>0.322097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Bexarotene</th>\n",
       "      <td>-0.489351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Dabrafenib</th>\n",
       "      <td>-0.529728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_darinaparsin</th>\n",
       "      <td>0.340990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_afatinib</th>\n",
       "      <td>-0.314642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Rucaparib</th>\n",
       "      <td>0.495838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_ibrutinib</th>\n",
       "      <td>-0.320376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_vandetanib</th>\n",
       "      <td>-0.211099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_sildenafil</th>\n",
       "      <td>0.331807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_SB-743921</th>\n",
       "      <td>0.219957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Bicalutamide</th>\n",
       "      <td>-0.368253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_MK-2206</th>\n",
       "      <td>-0.341257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_erismodegib</th>\n",
       "      <td>-0.317815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_OSI-027</th>\n",
       "      <td>0.252393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_valdecoxib</th>\n",
       "      <td>-0.274321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_mitomycin</th>\n",
       "      <td>0.278049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Fedratinib</th>\n",
       "      <td>-0.289089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_OSI-027</th>\n",
       "      <td>0.233698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_serdemetan</th>\n",
       "      <td>0.402967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_GSK461364</th>\n",
       "      <td>0.266745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_bleomycin A2</th>\n",
       "      <td>0.245910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ccle_Lapatinib</th>\n",
       "      <td>-0.277017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_bortezomib</th>\n",
       "      <td>0.219439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Dasatinib</th>\n",
       "      <td>-0.394111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Linsitinib</th>\n",
       "      <td>0.623647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_EX-527</th>\n",
       "      <td>0.224208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Selisistat</th>\n",
       "      <td>0.393335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Docetaxel</th>\n",
       "      <td>-0.315065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Belinostat</th>\n",
       "      <td>0.228368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_dinaciclib</th>\n",
       "      <td>0.216176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Dactolisib</th>\n",
       "      <td>-0.306925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_niclosamide</th>\n",
       "      <td>0.277728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_MK-1775</th>\n",
       "      <td>0.230026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_momelotinib</th>\n",
       "      <td>0.239119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Bosutinib</th>\n",
       "      <td>-0.471298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_pazopanib</th>\n",
       "      <td>0.290467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_pevonedistat</th>\n",
       "      <td>0.213294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_AT13387</th>\n",
       "      <td>-0.284561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Linifanib</th>\n",
       "      <td>-0.723077</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     DMB006\n",
       "ctrp_elocalcitol   0.345501\n",
       "ctrp_RAF265        0.211422\n",
       "gdsc_Vinblastine  -0.282946\n",
       "gdsc_Tamoxifen    -0.344953\n",
       "ctrp_temozolomide  0.273595\n",
       "ccle_Paclitaxel   -0.212470\n",
       "ctrp_quizartinib   0.269583\n",
       "gdsc_Sorafenib    -0.487374\n",
       "gdsc_BI-2536      -0.298340\n",
       "ctrp_daporinad     0.248200\n",
       "ctrp_MGCD-265      0.309477\n",
       "gdsc_Bortezomib   -0.373787\n",
       "ctrp_indisulam     0.390994\n",
       "ctrp_sunitinib     0.283824\n",
       "gdsc_Ponatinib    -0.337840\n",
       "gdsc_Methotrexate -0.264438\n",
       "ctrp_RO4929097     0.290939\n",
       "ctrp_TG-101348     0.238428\n",
       "gdsc_Axitinib     -0.489878\n",
       "ctrp_saracatinib  -0.252630\n",
       "ctrp_linsitinib    0.449036\n",
       "ctrp_etoposide     0.322097\n",
       "gdsc_Bexarotene   -0.489351\n",
       "gdsc_Dabrafenib   -0.529728\n",
       "ctrp_darinaparsin  0.340990\n",
       "ctrp_afatinib     -0.314642\n",
       "gdsc_Rucaparib     0.495838\n",
       "ctrp_ibrutinib    -0.320376\n",
       "ctrp_vandetanib   -0.211099\n",
       "ctrp_sildenafil    0.331807\n",
       "...                     ...\n",
       "ctrp_SB-743921     0.219957\n",
       "gdsc_Bicalutamide -0.368253\n",
       "gdsc_MK-2206      -0.341257\n",
       "ctrp_erismodegib  -0.317815\n",
       "ctrp_OSI-027       0.252393\n",
       "ctrp_valdecoxib   -0.274321\n",
       "ctrp_mitomycin     0.278049\n",
       "gdsc_Fedratinib   -0.289089\n",
       "gdsc_OSI-027       0.233698\n",
       "ctrp_serdemetan    0.402967\n",
       "ctrp_GSK461364     0.266745\n",
       "ctrp_bleomycin A2  0.245910\n",
       "ccle_Lapatinib    -0.277017\n",
       "ctrp_bortezomib    0.219439\n",
       "gdsc_Dasatinib    -0.394111\n",
       "gdsc_Linsitinib    0.623647\n",
       "ctrp_EX-527        0.224208\n",
       "gdsc_Selisistat    0.393335\n",
       "gdsc_Docetaxel    -0.315065\n",
       "gdsc_Belinostat    0.228368\n",
       "ctrp_dinaciclib    0.216176\n",
       "gdsc_Dactolisib   -0.306925\n",
       "ctrp_niclosamide   0.277728\n",
       "ctrp_MK-1775       0.230026\n",
       "ctrp_momelotinib   0.239119\n",
       "gdsc_Bosutinib    -0.471298\n",
       "ctrp_pazopanib     0.290467\n",
       "ctrp_pevonedistat  0.213294\n",
       "ctrp_AT13387      -0.284561\n",
       "gdsc_Linifanib    -0.723077\n",
       "\n",
       "[263 rows x 1 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from drug_suggestion.drug_annotation import subset_to_reasonable_drugs\n",
    "from drug_suggestion.expression.discover import load_discover_drug_to_cids\n",
    "disco2cid = load_discover_drug_to_cids()\n",
    "reasonable_results = subset_to_reasonable_drugs(setup.raw_discover_results, \n",
    "                                                disco2cid, \n",
    "                                                out_prefix='discover.{}'.format(setup.expression_control), \n",
    "                                                out_dir=setup.discover_out_dir)\n",
    "reasonable_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T21:38:37.402234Z",
     "start_time": "2018-12-14T21:38:36.965571Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/pandas/util/_decorators.py:118: FutureWarning: The `sheetname` keyword is deprecated, use `sheet_name` instead\n",
      "  return func(*args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "600 155\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>moa</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>drug</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gdsc_Navitoclax</th>\n",
       "      <td>0.647316</td>\n",
       "      <td>Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Linsitinib</th>\n",
       "      <td>0.623647</td>\n",
       "      <td>IGF-1R inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_TL-2-105</th>\n",
       "      <td>0.608240</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Tubastatin A</th>\n",
       "      <td>0.584972</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_SB52334</th>\n",
       "      <td>0.584363</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GSK1070916</th>\n",
       "      <td>0.565351</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GSK429286A</th>\n",
       "      <td>0.538728</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_QL-XII-61</th>\n",
       "      <td>0.531059</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_VX-702</th>\n",
       "      <td>0.526880</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GW-2580</th>\n",
       "      <td>0.506626</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_BX-912</th>\n",
       "      <td>0.497551</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Rucaparib</th>\n",
       "      <td>0.495838</td>\n",
       "      <td>PARP inhibitor, inhibits DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_linsitinib</th>\n",
       "      <td>0.449036</td>\n",
       "      <td>IGF-1R inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_staurosporine</th>\n",
       "      <td>0.445118</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_BRD-K99006945</th>\n",
       "      <td>0.444154</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_tubastatin A</th>\n",
       "      <td>0.442499</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_BRD-K34485477</th>\n",
       "      <td>0.434520</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GSK319347A</th>\n",
       "      <td>0.432345</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_dabrafenib</th>\n",
       "      <td>0.431378</td>\n",
       "      <td>Braf (V600E) inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_CCT036477</th>\n",
       "      <td>0.430151</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Amuvatinib</th>\n",
       "      <td>0.427398</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_CIL55</th>\n",
       "      <td>0.427059</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_NSC-87877</th>\n",
       "      <td>0.422465</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_isoliquiritigenin</th>\n",
       "      <td>0.418062</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_XMD13-2</th>\n",
       "      <td>0.416037</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_XL765</th>\n",
       "      <td>0.414014</td>\n",
       "      <td>PI3K/mTOR inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_avrainvillamide</th>\n",
       "      <td>0.413484</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_ABT-737</th>\n",
       "      <td>0.412971</td>\n",
       "      <td>Bcl-2/Bcl-xL inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_ML050</th>\n",
       "      <td>0.406183</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctrp_ML239</th>\n",
       "      <td>0.406129</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Pelitinib</th>\n",
       "      <td>-0.492027</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_XMD14-99</th>\n",
       "      <td>-0.497725</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Gefitinib</th>\n",
       "      <td>-0.502170</td>\n",
       "      <td>EGFR inhibitor (Iressa)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Y-39983</th>\n",
       "      <td>-0.505486</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_T0901317</th>\n",
       "      <td>-0.515939</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_QL-XI-92</th>\n",
       "      <td>-0.518282</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_AZ628</th>\n",
       "      <td>-0.522333</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Selumetinib</th>\n",
       "      <td>-0.523305</td>\n",
       "      <td>MEK1 inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_PD173074</th>\n",
       "      <td>-0.524387</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_rTRAIL</th>\n",
       "      <td>-0.526653</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Dabrafenib</th>\n",
       "      <td>-0.529728</td>\n",
       "      <td>Braf (V600E) inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Tivozanib</th>\n",
       "      <td>-0.531539</td>\n",
       "      <td>VEGFR inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_VNLG/124</th>\n",
       "      <td>-0.535228</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Afatinib</th>\n",
       "      <td>-0.538775</td>\n",
       "      <td>EGFR and ErbB2 inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GNF-2</th>\n",
       "      <td>-0.545472</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Idelalisib</th>\n",
       "      <td>-0.549082</td>\n",
       "      <td>Idelalisib,  PI3K/p110δ inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Cyclopamine</th>\n",
       "      <td>-0.552751</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Nutlin-3a (-)</th>\n",
       "      <td>-0.553077</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_WZ-1-84</th>\n",
       "      <td>-0.568081</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_GW441756</th>\n",
       "      <td>-0.571978</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_PHA-665752</th>\n",
       "      <td>-0.584666</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Quizartinib</th>\n",
       "      <td>-0.589566</td>\n",
       "      <td>Flt3 inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Cetuximab</th>\n",
       "      <td>-0.592119</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Ruxolitinib</th>\n",
       "      <td>-0.592466</td>\n",
       "      <td>JAK1/2 inhibitor, used for myelofibrosis and p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_VX-11e</th>\n",
       "      <td>-0.593326</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Crizotinib</th>\n",
       "      <td>-0.598021</td>\n",
       "      <td>ALK, ROS1 and MET inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Tretinoin</th>\n",
       "      <td>-0.614413</td>\n",
       "      <td>Retinoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Motesanib</th>\n",
       "      <td>-0.622760</td>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Alectinib</th>\n",
       "      <td>-0.679681</td>\n",
       "      <td>ALK inhibitor, used to treat non-small-cell lu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdsc_Linifanib</th>\n",
       "      <td>-0.723077</td>\n",
       "      <td>RTK inhibitor, targets VEGFR, PDGFR, KDR, and ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>755 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           score  \\\n",
       "drug                               \n",
       "gdsc_Navitoclax         0.647316   \n",
       "gdsc_Linsitinib         0.623647   \n",
       "gdsc_TL-2-105           0.608240   \n",
       "gdsc_Tubastatin A       0.584972   \n",
       "gdsc_SB52334            0.584363   \n",
       "gdsc_GSK1070916         0.565351   \n",
       "gdsc_GSK429286A         0.538728   \n",
       "gdsc_QL-XII-61          0.531059   \n",
       "gdsc_VX-702             0.526880   \n",
       "gdsc_GW-2580            0.506626   \n",
       "gdsc_BX-912             0.497551   \n",
       "gdsc_Rucaparib          0.495838   \n",
       "ctrp_linsitinib         0.449036   \n",
       "ctrp_staurosporine      0.445118   \n",
       "ctrp_BRD-K99006945      0.444154   \n",
       "ctrp_tubastatin A       0.442499   \n",
       "ctrp_BRD-K34485477      0.434520   \n",
       "gdsc_GSK319347A         0.432345   \n",
       "ctrp_dabrafenib         0.431378   \n",
       "ctrp_CCT036477          0.430151   \n",
       "gdsc_Amuvatinib         0.427398   \n",
       "ctrp_CIL55              0.427059   \n",
       "gdsc_NSC-87877          0.422465   \n",
       "ctrp_isoliquiritigenin  0.418062   \n",
       "gdsc_XMD13-2            0.416037   \n",
       "ctrp_XL765              0.414014   \n",
       "ctrp_avrainvillamide    0.413484   \n",
       "ctrp_ABT-737            0.412971   \n",
       "ctrp_ML050              0.406183   \n",
       "ctrp_ML239              0.406129   \n",
       "...                          ...   \n",
       "gdsc_Pelitinib         -0.492027   \n",
       "gdsc_XMD14-99          -0.497725   \n",
       "gdsc_Gefitinib         -0.502170   \n",
       "gdsc_Y-39983           -0.505486   \n",
       "gdsc_T0901317          -0.515939   \n",
       "gdsc_QL-XI-92          -0.518282   \n",
       "gdsc_AZ628             -0.522333   \n",
       "gdsc_Selumetinib       -0.523305   \n",
       "gdsc_PD173074          -0.524387   \n",
       "gdsc_rTRAIL            -0.526653   \n",
       "gdsc_Dabrafenib        -0.529728   \n",
       "gdsc_Tivozanib         -0.531539   \n",
       "gdsc_VNLG/124          -0.535228   \n",
       "gdsc_Afatinib          -0.538775   \n",
       "gdsc_GNF-2             -0.545472   \n",
       "gdsc_Idelalisib        -0.549082   \n",
       "gdsc_Cyclopamine       -0.552751   \n",
       "gdsc_Nutlin-3a (-)     -0.553077   \n",
       "gdsc_WZ-1-84           -0.568081   \n",
       "gdsc_GW441756          -0.571978   \n",
       "gdsc_PHA-665752        -0.584666   \n",
       "gdsc_Quizartinib       -0.589566   \n",
       "gdsc_Cetuximab         -0.592119   \n",
       "gdsc_Ruxolitinib       -0.592466   \n",
       "gdsc_VX-11e            -0.593326   \n",
       "gdsc_Crizotinib        -0.598021   \n",
       "gdsc_Tretinoin         -0.614413   \n",
       "gdsc_Motesanib         -0.622760   \n",
       "gdsc_Alectinib         -0.679681   \n",
       "gdsc_Linifanib         -0.723077   \n",
       "\n",
       "                                                                      moa  \n",
       "drug                                                                       \n",
       "gdsc_Navitoclax         Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...  \n",
       "gdsc_Linsitinib                                          IGF-1R inhibitor  \n",
       "gdsc_TL-2-105                                     Not Clinically Relevant  \n",
       "gdsc_Tubastatin A                                 Not Clinically Relevant  \n",
       "gdsc_SB52334                                      Not Clinically Relevant  \n",
       "gdsc_GSK1070916                                   Not Clinically Relevant  \n",
       "gdsc_GSK429286A                                   Not Clinically Relevant  \n",
       "gdsc_QL-XII-61                                    Not Clinically Relevant  \n",
       "gdsc_VX-702                                       Not Clinically Relevant  \n",
       "gdsc_GW-2580                                      Not Clinically Relevant  \n",
       "gdsc_BX-912                                       Not Clinically Relevant  \n",
       "gdsc_Rucaparib                        PARP inhibitor, inhibits DNA repair  \n",
       "ctrp_linsitinib                                          IGF-1R inhibitor  \n",
       "ctrp_staurosporine                                Not Clinically Relevant  \n",
       "ctrp_BRD-K99006945                                Not Clinically Relevant  \n",
       "ctrp_tubastatin A                                 Not Clinically Relevant  \n",
       "ctrp_BRD-K34485477                                Not Clinically Relevant  \n",
       "gdsc_GSK319347A                                   Not Clinically Relevant  \n",
       "ctrp_dabrafenib                                    Braf (V600E) inhibitor  \n",
       "ctrp_CCT036477                                    Not Clinically Relevant  \n",
       "gdsc_Amuvatinib                                   Not Clinically Relevant  \n",
       "ctrp_CIL55                                        Not Clinically Relevant  \n",
       "gdsc_NSC-87877                                    Not Clinically Relevant  \n",
       "ctrp_isoliquiritigenin                            Not Clinically Relevant  \n",
       "gdsc_XMD13-2                                      Not Clinically Relevant  \n",
       "ctrp_XL765                                            PI3K/mTOR inhibitor  \n",
       "ctrp_avrainvillamide                              Not Clinically Relevant  \n",
       "ctrp_ABT-737                                       Bcl-2/Bcl-xL inhibitor  \n",
       "ctrp_ML050                                        Not Clinically Relevant  \n",
       "ctrp_ML239                                        Not Clinically Relevant  \n",
       "...                                                                   ...  \n",
       "gdsc_Pelitinib                                    Not Clinically Relevant  \n",
       "gdsc_XMD14-99                                     Not Clinically Relevant  \n",
       "gdsc_Gefitinib                                    EGFR inhibitor (Iressa)  \n",
       "gdsc_Y-39983                                      Not Clinically Relevant  \n",
       "gdsc_T0901317                                     Not Clinically Relevant  \n",
       "gdsc_QL-XI-92                                     Not Clinically Relevant  \n",
       "gdsc_AZ628                                        Not Clinically Relevant  \n",
       "gdsc_Selumetinib                                           MEK1 inhibitor  \n",
       "gdsc_PD173074                                     Not Clinically Relevant  \n",
       "gdsc_rTRAIL                                       Not Clinically Relevant  \n",
       "gdsc_Dabrafenib                                    Braf (V600E) inhibitor  \n",
       "gdsc_Tivozanib                                            VEGFR inhibitor  \n",
       "gdsc_VNLG/124                                     Not Clinically Relevant  \n",
       "gdsc_Afatinib                                    EGFR and ErbB2 inhibitor  \n",
       "gdsc_GNF-2                                        Not Clinically Relevant  \n",
       "gdsc_Idelalisib                         Idelalisib,  PI3K/p110δ inhibitor  \n",
       "gdsc_Cyclopamine                                  Not Clinically Relevant  \n",
       "gdsc_Nutlin-3a (-)                                Not Clinically Relevant  \n",
       "gdsc_WZ-1-84                                      Not Clinically Relevant  \n",
       "gdsc_GW441756                                     Not Clinically Relevant  \n",
       "gdsc_PHA-665752                                   Not Clinically Relevant  \n",
       "gdsc_Quizartinib                                           Flt3 inhibitor  \n",
       "gdsc_Cetuximab                                    Not Clinically Relevant  \n",
       "gdsc_Ruxolitinib        JAK1/2 inhibitor, used for myelofibrosis and p...  \n",
       "gdsc_VX-11e                                       Not Clinically Relevant  \n",
       "gdsc_Crizotinib                               ALK, ROS1 and MET inhibitor  \n",
       "gdsc_Tretinoin                                                   Retinoid  \n",
       "gdsc_Motesanib                                    Not Clinically Relevant  \n",
       "gdsc_Alectinib          ALK inhibitor, used to treat non-small-cell lu...  \n",
       "gdsc_Linifanib          RTK inhibitor, targets VEGFR, PDGFR, KDR, and ...  \n",
       "\n",
       "[755 rows x 2 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from companion_script import format_drugs\n",
    "from drug_suggestion.expression.discover import load_discover_drug_to_cids\n",
    "disco2cid = load_discover_drug_to_cids()\n",
    "reasonable_results = format_drugs(setup.raw_discover_results, \n",
    "                                                disco2cid, \n",
    "                                                out_prefix='discover.{}'.format(setup.expression_control), \n",
    "                                                out_dir=setup.discover_out_dir)\n",
    "reasonable_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-14T22:06:26.544611Z",
     "start_time": "2018-12-14T22:06:26.470364Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>moa</th>\n",
       "      <th>GDSC</th>\n",
       "      <th>CTRP</th>\n",
       "      <th>CCLE</th>\n",
       "      <th>drug</th>\n",
       "      <th>score</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>tl-2-105</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.608</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_TL-2-105</td>\n",
       "      <td>0.608</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sb52334</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.584</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_SB52334</td>\n",
       "      <td>0.584</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gsk1070916</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.565</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GSK1070916</td>\n",
       "      <td>0.565</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gsk429286a</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GSK429286A</td>\n",
       "      <td>0.539</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>linsitinib</th>\n",
       "      <td>IGF-1R inhibitor</td>\n",
       "      <td>0.624</td>\n",
       "      <td>0.449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_linsitinib</td>\n",
       "      <td>0.536</td>\n",
       "      <td>++.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ql-xii-61</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_QL-XII-61</td>\n",
       "      <td>0.531</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vx-702</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.527</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_VX-702</td>\n",
       "      <td>0.527</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tubastatin a</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.585</td>\n",
       "      <td>0.442</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_tubastatin A</td>\n",
       "      <td>0.514</td>\n",
       "      <td>++.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gw-2580</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.507</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GW-2580</td>\n",
       "      <td>0.507</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bx-912</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_BX-912</td>\n",
       "      <td>0.498</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rucaparib</th>\n",
       "      <td>PARP inhibitor, inhibits DNA repair</td>\n",
       "      <td>0.496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Rucaparib</td>\n",
       "      <td>0.496</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>navitoclax</th>\n",
       "      <td>Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...</td>\n",
       "      <td>0.647</td>\n",
       "      <td>0.307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_navitoclax</td>\n",
       "      <td>0.477</td>\n",
       "      <td>++.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>staurosporine</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_staurosporine</td>\n",
       "      <td>0.445</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brd-k99006945</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_BRD-K99006945</td>\n",
       "      <td>0.444</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brd-k34485477</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_BRD-K34485477</td>\n",
       "      <td>0.435</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gsk319347a</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.432</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GSK319347A</td>\n",
       "      <td>0.432</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cct036477</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_CCT036477</td>\n",
       "      <td>0.430</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amuvatinib</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Amuvatinib</td>\n",
       "      <td>0.427</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cil55</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_CIL55</td>\n",
       "      <td>0.427</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nsc-87877</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_NSC-87877</td>\n",
       "      <td>0.422</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoliquiritigenin</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_isoliquiritigenin</td>\n",
       "      <td>0.418</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xmd13-2</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>0.416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_XMD13-2</td>\n",
       "      <td>0.416</td>\n",
       "      <td>+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xl765</th>\n",
       "      <td>PI3K/mTOR inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.414</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_XL765</td>\n",
       "      <td>0.414</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>avrainvillamide</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_avrainvillamide</td>\n",
       "      <td>0.413</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abt-737</th>\n",
       "      <td>Bcl-2/Bcl-xL inhibitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_ABT-737</td>\n",
       "      <td>0.413</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ml050</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_ML050</td>\n",
       "      <td>0.406</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ml239</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_ML239</td>\n",
       "      <td>0.406</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brd-k88742110</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_BRD-K88742110</td>\n",
       "      <td>0.401</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brd-k26531177</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_BRD-K26531177</td>\n",
       "      <td>0.398</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>skepinone-l</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ctrp_skepinone-L</td>\n",
       "      <td>0.395</td>\n",
       "      <td>.+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bryostatin 1</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Bryostatin 1</td>\n",
       "      <td>-0.436</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ro-3306</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_RO-3306</td>\n",
       "      <td>-0.437</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tak-715</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_TAK-715</td>\n",
       "      <td>-0.440</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z-llnle-cho</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.442</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Z-LLNle-CHO</td>\n",
       "      <td>-0.442</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cp724714</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_CP724714</td>\n",
       "      <td>-0.450</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tozasertib</th>\n",
       "      <td>VX-680, pan-Aurora inhibitor, mostly against A...</td>\n",
       "      <td>-0.464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Tozasertib</td>\n",
       "      <td>-0.464</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kin001-260</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_KIN001-260</td>\n",
       "      <td>-0.466</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ruxolitinib</th>\n",
       "      <td>JAK1/2 inhibitor, used for myelofibrosis and p...</td>\n",
       "      <td>-0.592</td>\n",
       "      <td>-0.346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Ruxolitinib</td>\n",
       "      <td>-0.469</td>\n",
       "      <td>--.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qs11</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_QS11</td>\n",
       "      <td>-0.484</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rapamycin</th>\n",
       "      <td>rapamycin, mTOR inhibitor</td>\n",
       "      <td>-0.487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Rapamycin</td>\n",
       "      <td>-0.487</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pelitinib</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.492</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Pelitinib</td>\n",
       "      <td>-0.492</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xmd14-99</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_XMD14-99</td>\n",
       "      <td>-0.498</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y-39983</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Y-39983</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>linifanib</th>\n",
       "      <td>RTK inhibitor, targets VEGFR, PDGFR, KDR, and ...</td>\n",
       "      <td>-0.723</td>\n",
       "      <td>-0.292</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Linifanib</td>\n",
       "      <td>-0.507</td>\n",
       "      <td>--.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t0901317</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_T0901317</td>\n",
       "      <td>-0.516</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ql-xi-92</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_QL-XI-92</td>\n",
       "      <td>-0.518</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>az628</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.522</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_AZ628</td>\n",
       "      <td>-0.522</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pd173074</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_PD173074</td>\n",
       "      <td>-0.524</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rtrail</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.527</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_rTRAIL</td>\n",
       "      <td>-0.527</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vnlg/124</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_VNLG/124</td>\n",
       "      <td>-0.535</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gnf-2</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GNF-2</td>\n",
       "      <td>-0.545</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>idelalisib</th>\n",
       "      <td>Idelalisib,  PI3K/p110δ inhibitor</td>\n",
       "      <td>-0.549</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Idelalisib</td>\n",
       "      <td>-0.549</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyclopamine</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Cyclopamine</td>\n",
       "      <td>-0.553</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nutlin-3a (-)</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Nutlin-3a (-)</td>\n",
       "      <td>-0.553</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wz-1-84</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_WZ-1-84</td>\n",
       "      <td>-0.568</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gw441756</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_GW441756</td>\n",
       "      <td>-0.572</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cetuximab</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.592</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Cetuximab</td>\n",
       "      <td>-0.592</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vx-11e</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.593</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_VX-11e</td>\n",
       "      <td>-0.593</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>motesanib</th>\n",
       "      <td>Not Clinically Relevant</td>\n",
       "      <td>-0.623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Motesanib</td>\n",
       "      <td>-0.623</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alectinib</th>\n",
       "      <td>ALK inhibitor, used to treat non-small-cell lu...</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gdsc_Alectinib</td>\n",
       "      <td>-0.680</td>\n",
       "      <td>-..</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>672 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 moa   GDSC  \\\n",
       "tl-2-105                                     Not Clinically Relevant  0.608   \n",
       "sb52334                                      Not Clinically Relevant  0.584   \n",
       "gsk1070916                                   Not Clinically Relevant  0.565   \n",
       "gsk429286a                                   Not Clinically Relevant  0.539   \n",
       "linsitinib                                          IGF-1R inhibitor  0.624   \n",
       "ql-xii-61                                    Not Clinically Relevant  0.531   \n",
       "vx-702                                       Not Clinically Relevant  0.527   \n",
       "tubastatin a                                 Not Clinically Relevant  0.585   \n",
       "gw-2580                                      Not Clinically Relevant  0.507   \n",
       "bx-912                                       Not Clinically Relevant  0.498   \n",
       "rucaparib                        PARP inhibitor, inhibits DNA repair  0.496   \n",
       "navitoclax         Bcl-2 family inhibitor: esp Bcl-xL, Bcl-2 and ...  0.647   \n",
       "staurosporine                                Not Clinically Relevant    NaN   \n",
       "brd-k99006945                                Not Clinically Relevant    NaN   \n",
       "brd-k34485477                                Not Clinically Relevant    NaN   \n",
       "gsk319347a                                   Not Clinically Relevant  0.432   \n",
       "cct036477                                    Not Clinically Relevant    NaN   \n",
       "amuvatinib                                   Not Clinically Relevant  0.427   \n",
       "cil55                                        Not Clinically Relevant    NaN   \n",
       "nsc-87877                                    Not Clinically Relevant  0.422   \n",
       "isoliquiritigenin                            Not Clinically Relevant    NaN   \n",
       "xmd13-2                                      Not Clinically Relevant  0.416   \n",
       "xl765                                            PI3K/mTOR inhibitor    NaN   \n",
       "avrainvillamide                              Not Clinically Relevant    NaN   \n",
       "abt-737                                       Bcl-2/Bcl-xL inhibitor    NaN   \n",
       "ml050                                        Not Clinically Relevant    NaN   \n",
       "ml239                                        Not Clinically Relevant    NaN   \n",
       "brd-k88742110                                Not Clinically Relevant    NaN   \n",
       "brd-k26531177                                Not Clinically Relevant    NaN   \n",
       "skepinone-l                                  Not Clinically Relevant    NaN   \n",
       "...                                                              ...    ...   \n",
       "bryostatin 1                                 Not Clinically Relevant -0.436   \n",
       "ro-3306                                      Not Clinically Relevant -0.437   \n",
       "tak-715                                      Not Clinically Relevant  -0.44   \n",
       "z-llnle-cho                                  Not Clinically Relevant -0.442   \n",
       "cp724714                                     Not Clinically Relevant  -0.45   \n",
       "tozasertib         VX-680, pan-Aurora inhibitor, mostly against A... -0.464   \n",
       "kin001-260                                   Not Clinically Relevant -0.466   \n",
       "ruxolitinib        JAK1/2 inhibitor, used for myelofibrosis and p... -0.592   \n",
       "qs11                                         Not Clinically Relevant -0.484   \n",
       "rapamycin                                  rapamycin, mTOR inhibitor -0.487   \n",
       "pelitinib                                    Not Clinically Relevant -0.492   \n",
       "xmd14-99                                     Not Clinically Relevant -0.498   \n",
       "y-39983                                      Not Clinically Relevant -0.505   \n",
       "linifanib          RTK inhibitor, targets VEGFR, PDGFR, KDR, and ... -0.723   \n",
       "t0901317                                     Not Clinically Relevant -0.516   \n",
       "ql-xi-92                                     Not Clinically Relevant -0.518   \n",
       "az628                                        Not Clinically Relevant -0.522   \n",
       "pd173074                                     Not Clinically Relevant -0.524   \n",
       "rtrail                                       Not Clinically Relevant -0.527   \n",
       "vnlg/124                                     Not Clinically Relevant -0.535   \n",
       "gnf-2                                        Not Clinically Relevant -0.545   \n",
       "idelalisib                         Idelalisib,  PI3K/p110δ inhibitor -0.549   \n",
       "cyclopamine                                  Not Clinically Relevant -0.553   \n",
       "nutlin-3a (-)                                Not Clinically Relevant -0.553   \n",
       "wz-1-84                                      Not Clinically Relevant -0.568   \n",
       "gw441756                                     Not Clinically Relevant -0.572   \n",
       "cetuximab                                    Not Clinically Relevant -0.592   \n",
       "vx-11e                                       Not Clinically Relevant -0.593   \n",
       "motesanib                                    Not Clinically Relevant -0.623   \n",
       "alectinib          ALK inhibitor, used to treat non-small-cell lu...  -0.68   \n",
       "\n",
       "                    CTRP CCLE                    drug  score evidence  \n",
       "tl-2-105             NaN  NaN           gdsc_TL-2-105  0.608      +..  \n",
       "sb52334              NaN  NaN            gdsc_SB52334  0.584      +..  \n",
       "gsk1070916           NaN  NaN         gdsc_GSK1070916  0.565      +..  \n",
       "gsk429286a           NaN  NaN         gdsc_GSK429286A  0.539      +..  \n",
       "linsitinib         0.449  NaN         ctrp_linsitinib  0.536      ++.  \n",
       "ql-xii-61            NaN  NaN          gdsc_QL-XII-61  0.531      +..  \n",
       "vx-702               NaN  NaN             gdsc_VX-702  0.527      +..  \n",
       "tubastatin a       0.442  NaN       ctrp_tubastatin A  0.514      ++.  \n",
       "gw-2580              NaN  NaN            gdsc_GW-2580  0.507      +..  \n",
       "bx-912               NaN  NaN             gdsc_BX-912  0.498      +..  \n",
       "rucaparib            NaN  NaN          gdsc_Rucaparib  0.496      +..  \n",
       "navitoclax         0.307  NaN         ctrp_navitoclax  0.477      ++.  \n",
       "staurosporine      0.445  NaN      ctrp_staurosporine  0.445      .+.  \n",
       "brd-k99006945      0.444  NaN      ctrp_BRD-K99006945  0.444      .+.  \n",
       "brd-k34485477      0.435  NaN      ctrp_BRD-K34485477  0.435      .+.  \n",
       "gsk319347a           NaN  NaN         gdsc_GSK319347A  0.432      +..  \n",
       "cct036477           0.43  NaN          ctrp_CCT036477  0.430      .+.  \n",
       "amuvatinib           NaN  NaN         gdsc_Amuvatinib  0.427      +..  \n",
       "cil55              0.427  NaN              ctrp_CIL55  0.427      .+.  \n",
       "nsc-87877            NaN  NaN          gdsc_NSC-87877  0.422      +..  \n",
       "isoliquiritigenin  0.418  NaN  ctrp_isoliquiritigenin  0.418      .+.  \n",
       "xmd13-2              NaN  NaN            gdsc_XMD13-2  0.416      +..  \n",
       "xl765              0.414  NaN              ctrp_XL765  0.414      .+.  \n",
       "avrainvillamide    0.413  NaN    ctrp_avrainvillamide  0.413      .+.  \n",
       "abt-737            0.413  NaN            ctrp_ABT-737  0.413      .+.  \n",
       "ml050              0.406  NaN              ctrp_ML050  0.406      .+.  \n",
       "ml239              0.406  NaN              ctrp_ML239  0.406      .+.  \n",
       "brd-k88742110      0.401  NaN      ctrp_BRD-K88742110  0.401      .+.  \n",
       "brd-k26531177      0.398  NaN      ctrp_BRD-K26531177  0.398      .+.  \n",
       "skepinone-l        0.395  NaN        ctrp_skepinone-L  0.395      .+.  \n",
       "...                  ...  ...                     ...    ...      ...  \n",
       "bryostatin 1         NaN  NaN       gdsc_Bryostatin 1 -0.436      -..  \n",
       "ro-3306              NaN  NaN            gdsc_RO-3306 -0.437      -..  \n",
       "tak-715              NaN  NaN            gdsc_TAK-715 -0.440      -..  \n",
       "z-llnle-cho          NaN  NaN        gdsc_Z-LLNle-CHO -0.442      -..  \n",
       "cp724714             NaN  NaN           gdsc_CP724714 -0.450      -..  \n",
       "tozasertib           NaN  NaN         gdsc_Tozasertib -0.464      -..  \n",
       "kin001-260           NaN  NaN         gdsc_KIN001-260 -0.466      -..  \n",
       "ruxolitinib       -0.346  NaN        gdsc_Ruxolitinib -0.469      --.  \n",
       "qs11                 NaN  NaN               gdsc_QS11 -0.484      -..  \n",
       "rapamycin            NaN  NaN          gdsc_Rapamycin -0.487      -..  \n",
       "pelitinib            NaN  NaN          gdsc_Pelitinib -0.492      -..  \n",
       "xmd14-99             NaN  NaN           gdsc_XMD14-99 -0.498      -..  \n",
       "y-39983              NaN  NaN            gdsc_Y-39983 -0.505      -..  \n",
       "linifanib         -0.292  NaN          gdsc_Linifanib -0.507      --.  \n",
       "t0901317             NaN  NaN           gdsc_T0901317 -0.516      -..  \n",
       "ql-xi-92             NaN  NaN           gdsc_QL-XI-92 -0.518      -..  \n",
       "az628                NaN  NaN              gdsc_AZ628 -0.522      -..  \n",
       "pd173074             NaN  NaN           gdsc_PD173074 -0.524      -..  \n",
       "rtrail               NaN  NaN             gdsc_rTRAIL -0.527      -..  \n",
       "vnlg/124             NaN  NaN           gdsc_VNLG/124 -0.535      -..  \n",
       "gnf-2                NaN  NaN              gdsc_GNF-2 -0.545      -..  \n",
       "idelalisib           NaN  NaN         gdsc_Idelalisib -0.549      -..  \n",
       "cyclopamine          NaN  NaN        gdsc_Cyclopamine -0.553      -..  \n",
       "nutlin-3a (-)        NaN  NaN      gdsc_Nutlin-3a (-) -0.553      -..  \n",
       "wz-1-84              NaN  NaN            gdsc_WZ-1-84 -0.568      -..  \n",
       "gw441756             NaN  NaN           gdsc_GW441756 -0.572      -..  \n",
       "cetuximab            NaN  NaN          gdsc_Cetuximab -0.592      -..  \n",
       "vx-11e               NaN  NaN             gdsc_VX-11e -0.593      -..  \n",
       "motesanib            NaN  NaN          gdsc_Motesanib -0.623      -..  \n",
       "alectinib            NaN  NaN          gdsc_Alectinib -0.680      -..  \n",
       "\n",
       "[672 rows x 7 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setup.df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Select signature genes for Connectivity Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.669039Z",
     "start_time": "2018-12-17T15:54:04.613148Z"
    },
    "nbtools": {
     "description": "This function parses CMap's results.",
     "name": "make_cmap_slide",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d00a855db39405b87c91006652bfe1d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description=\"This function parses CMap's results.\", function_import='make_cmap_slide', name='make_cm…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(\n",
    "  description=\"This function parses CMap's results.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def make_cmap_slide(setup):\n",
    "    log('About to parse CMap results.')\n",
    "    patient_exp = pd.read_csv(setup.patient_gexp_file, index_col=0)\n",
    "    control_exp = load_control_exp(setup.expression_control)\n",
    "    cmap_genesets = make_cmap_genesets(patient_exp, control_exp)\n",
    "    write_cmap_genesets(cmap_genesets, setup.cmap_out_dir)\n",
    "\n",
    "    # must match path to downloaded .gct file\n",
    "    cmap_gct = os.path.join(setup.cmap_out_dir, 'cmap_result.gct')\n",
    "\n",
    "    if os.path.exists(cmap_gct):\n",
    "        cmap_ranked_drugs = read_cmap_gct(cmap_gct)\n",
    "        cmap_ranked_drugs.columns = [setup.case_id]\n",
    "        cmap_ranked_drugs.to_csv(setup.cmap_all_ranked_drugs_file)\n",
    "        cmap2cid = load_cmap_drug_to_cids()\n",
    "        cmap_reasonable = subset_to_reasonable_drugs(cmap_ranked_drugs.T, \n",
    "                                   cmap2cid,\n",
    "                                   out_prefix='cmap.{}'.format(setup.expression_control), \n",
    "                                   out_dir=setup.cmap_out_dir).sort_values(by=setup.case_id, ascending=False)\n",
    "        rdrugs_cmap = pd.read_csv(setup.cmap_reasonable_ranked_drugs_file, index_col=None)\n",
    "        make_exp_drug_ranking_results_slide(setup.mtb_ppt_file, rdrugs_cmap, setup.expression_control, method='CMap')\n",
    "        setup.rdrugs_cmap = rdrugs_cmap\n",
    "        pickle.dump(setup, file=open('patients/'+setup.patient_dir+'/drug_suggestions/'+setup.case_id+'_drug_recommendations.p','wb'))\n",
    "        log(\"done!\")\n",
    "    else:\n",
    "        log(f\"cmap_result.gct not found! (It should be present in the directiory {setup.cmap_out_dir}).\")\n",
    "        log(\"Try again if you'd like to see CMap results.\")\n",
    "        log(\"Hint, you may want to go here:\")\n",
    "        log(\"https://clue.io/l1000-query#individual\")\n",
    "    \n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup5_CMap.p','wb'))\n",
    "    pickle.dump(setup, file=open(os.path.join(setup.out_dir, setup.case_id+'_backup5_CMap.p'),'wb'))\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DiSCoVER \"supporting information\" or \"evidence\" means in how many of the three drug databases that drug is significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-17T15:54:04.723451Z",
     "start_time": "2018-12-17T15:54:04.673658Z"
    },
    "nbtools": {
     "description": "This function merges the results of DiSCoVER and CMap.",
     "name": "merge_discover_and_cmap",
     "param_values": {
      "output_var": "setup",
      "setup": "setup"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d2f6c6e483104e259fb55386f9f88048",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "UIBuilder(description='This function merges the results of DiSCoVER and CMap.', function_import='merge_discove…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@genepattern.build_ui(\n",
    "  description=\"This function merges the results of DiSCoVER and CMap.\",\n",
    "  parameters={\n",
    "    \"setup\": {\"default\": \"setup\",\n",
    "              \"hide\": False,\n",
    "              \"description\": \"The variable which has the setup information\"},\n",
    "    \"output_var\": {\n",
    "        \"default\": \"setup\",\n",
    "        \"hide\": True,\n",
    "    },\n",
    "})\n",
    "def merge_discover_and_cmap(setup):\n",
    "    log('Merging results from DiSCoVER and CMap.')\n",
    "    try:\n",
    "        combined_df = add_cmap_to_split_df(discover=setup.df,cmap=setup.rdrugs_cmap)\n",
    "    except AttributeError:\n",
    "        log(\"CMap info not present, proceeding with only DiSCoVER\")\n",
    "        combined_df = setup.df\n",
    "    to_slide = rank_combined_df(combined_df)\n",
    "    make_intersection_slide(setup.mtb_ppt_file, to_slide, setup.expression_control, method='DiSCoVER ∩ CMap')\n",
    "    \n",
    "    log(\"Done Merging results from DiSCoVER and CMap!\")\n",
    "    log(\"Saving combined_df and to_slide on setup variable\")\n",
    "    setup.combined_df = combined_df\n",
    "    setup.to_slide = to_slide\n",
    "    pickle.dump(setup, file=open('patients/'+setup.patient_dir+'/drug_suggestions/'+setup.case_id+'_drug_recommendations_merged.p','wb'))\n",
    "#     pickle.dump(setup, file=open(setup.out_dir+'_backup6_merged.p','wb'))\n",
    "    pickle.dump(setup, file=open(os.path.join(setup.out_dir, setup.case_id+'_backup6_merged.p'),'wb'))\n",
    "    \n",
    "    log(\"Done done!\")\n",
    "    return setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
